// qase-javascript-commons/dist/env/env-enum.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EnvLoggingEnum = exports.EnvLocalEnum = exports.EnvConfigurationsEnum = exports.EnvBatchEnum = exports.EnvPlanEnum = exports.EnvRunEnum = exports.EnvApiEnum = exports.EnvTestOpsEnum = exports.EnvEnum = void 0;
/**
 * @enum {string}
 */
var EnvEnum;
(function (EnvEnum) {
    EnvEnum["mode"] = "QASE_MODE";
    EnvEnum["fallback"] = "QASE_FALLBACK";
    EnvEnum["debug"] = "QASE_DEBUG";
    EnvEnum["environment"] = "QASE_ENVIRONMENT";
    EnvEnum["captureLogs"] = "QASE_CAPTURE_LOGS";
    EnvEnum["rootSuite"] = "QASE_ROOT_SUITE";
    EnvEnum["statusMapping"] = "QASE_STATUS_MAPPING";
})(EnvEnum || (exports.EnvEnum = EnvEnum = {}));
/**
 * @enum {string}
 */
var EnvTestOpsEnum;
(function (EnvTestOpsEnum) {
    EnvTestOpsEnum["project"] = "QASE_TESTOPS_PROJECT";
    EnvTestOpsEnum["uploadAttachments"] = "QASE_TESTOPS_UPLOAD_ATTACHMENTS";
    EnvTestOpsEnum["defect"] = "QASE_TESTOPS_DEFECT";
    EnvTestOpsEnum["statusFilter"] = "QASE_TESTOPS_STATUS_FILTER";
    EnvTestOpsEnum["showPublicReportLink"] = "QASE_TESTOPS_SHOW_PUBLIC_REPORT_LINK";
})(EnvTestOpsEnum || (exports.EnvTestOpsEnum = EnvTestOpsEnum = {}));
/**
 * @enum {string}
 */
var EnvApiEnum;
(function (EnvApiEnum) {
    EnvApiEnum["token"] = "QASE_TESTOPS_API_TOKEN";
    EnvApiEnum["host"] = "QASE_TESTOPS_API_HOST";
})(EnvApiEnum || (exports.EnvApiEnum = EnvApiEnum = {}));
/**
 * @enum {string}
 */
var EnvRunEnum;
(function (EnvRunEnum) {
    EnvRunEnum["id"] = "QASE_TESTOPS_RUN_ID";
    EnvRunEnum["title"] = "QASE_TESTOPS_RUN_TITLE";
    EnvRunEnum["description"] = "QASE_TESTOPS_RUN_DESCRIPTION";
    EnvRunEnum["complete"] = "QASE_TESTOPS_RUN_COMPLETE";
    EnvRunEnum["tags"] = "QASE_TESTOPS_RUN_TAGS";
    EnvRunEnum["externalLink"] = "QASE_TESTOPS_RUN_EXTERNAL_LINK";
})(EnvRunEnum || (exports.EnvRunEnum = EnvRunEnum = {}));
/**
 * @enum {string}
 */
var EnvPlanEnum;
(function (EnvPlanEnum) {
    EnvPlanEnum["id"] = "QASE_TESTOPS_PLAN_ID";
})(EnvPlanEnum || (exports.EnvPlanEnum = EnvPlanEnum = {}));
/**
 * @enum {string}
 */
var EnvBatchEnum;
(function (EnvBatchEnum) {
    EnvBatchEnum["size"] = "QASE_TESTOPS_BATCH_SIZE";
})(EnvBatchEnum || (exports.EnvBatchEnum = EnvBatchEnum = {}));
/**
 * @enum {string}
 */
var EnvConfigurationsEnum;
(function (EnvConfigurationsEnum) {
    EnvConfigurationsEnum["values"] = "QASE_TESTOPS_CONFIGURATIONS_VALUES";
    EnvConfigurationsEnum["createIfNotExists"] = "QASE_TESTOPS_CONFIGURATIONS_CREATE_IF_NOT_EXISTS";
})(EnvConfigurationsEnum || (exports.EnvConfigurationsEnum = EnvConfigurationsEnum = {}));
/**
 * @enum {string}
 */
var EnvLocalEnum;
(function (EnvLocalEnum) {
    EnvLocalEnum["path"] = "QASE_REPORT_CONNECTION_PATH";
    EnvLocalEnum["format"] = "QASE_REPORT_CONNECTION_FORMAT";
})(EnvLocalEnum || (exports.EnvLocalEnum = EnvLocalEnum = {}));
/**
 * @enum {string}
 */
var EnvLoggingEnum;
(function (EnvLoggingEnum) {
    EnvLoggingEnum["console"] = "QASE_LOGGING_CONSOLE";
    EnvLoggingEnum["file"] = "QASE_LOGGING_FILE";
})(EnvLoggingEnum || (exports.EnvLoggingEnum = EnvLoggingEnum = {}));
