// qase-javascript-commons/dist/models/test-result.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TestResultType = void 0;
const test_execution_1 = require("./test-execution");
class TestResultType {
    id;
    title;
    signature;
    run_id;
    testops_id;
    /**
     * Multi-project mapping: project code -> array of test case IDs.
     * When set, overrides testops_id for multi-project mode.
     * If empty/null, fall back to testops_id for single project.
     */
    testops_project_mapping;
    execution;
    fields;
    attachments;
    steps;
    params;
    group_params;
    author;
    relations;
    muted;
    message;
    preparedAttachments;
    constructor(title) {
        this.id = '';
        this.title = title;
        this.signature = '';
        this.run_id = null;
        this.testops_id = null;
        this.testops_project_mapping = null;
        this.execution = new test_execution_1.TestExecution();
        this.fields = {};
        this.attachments = [];
        this.steps = [];
        this.params = {};
        this.group_params = {};
        this.author = null;
        this.relations = null;
        this.muted = false;
        this.message = null;
        this.preparedAttachments = [];
    }
    /**
     * Set test case IDs for a specific project in multi-project mapping.
     */
    setTestopsProjectMapping(projectCode, testopsIds) {
        if (!this.testops_project_mapping) {
            this.testops_project_mapping = {};
        }
        this.testops_project_mapping[projectCode] = testopsIds;
    }
    /**
     * Get the entire project-to-IDs mapping.
     */
    getTestopsProjectMapping() {
        return this.testops_project_mapping;
    }
    /**
     * Get test case IDs for a specific project.
     */
    getTestopsIdsForProject(projectCode) {
        return this.testops_project_mapping?.[projectCode];
    }
    /**
     * Get list of all project codes in the mapping.
     */
    getProjects() {
        return this.testops_project_mapping ? Object.keys(this.testops_project_mapping) : [];
    }
}
exports.TestResultType = TestResultType;
