// qase-api-client/dist/model/requirement-query.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Qase.io TestOps API v1
 * Qase TestOps API v1 Specification.
 *
 * The version of the OpenAPI document: 1.0.0
 * Contact: support@qase.io
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.RequirementQueryTypeEnum = exports.RequirementQueryStatusEnum = void 0;
/**
    * @export
    * @enum {string}
    */
var RequirementQueryStatusEnum;
(function (RequirementQueryStatusEnum) {
    RequirementQueryStatusEnum["VALID"] = "valid";
    RequirementQueryStatusEnum["DRAFT"] = "draft";
    RequirementQueryStatusEnum["REVIEW"] = "review";
    RequirementQueryStatusEnum["REWORK"] = "rework";
    RequirementQueryStatusEnum["FINISH"] = "finish";
    RequirementQueryStatusEnum["IMPLEMENTED"] = "implemented";
    RequirementQueryStatusEnum["NOT_TESTABLE"] = "not-testable";
    RequirementQueryStatusEnum["OBSOLETE"] = "obsolete";
})(RequirementQueryStatusEnum || (exports.RequirementQueryStatusEnum = RequirementQueryStatusEnum = {}));
/**
    * @export
    * @enum {string}
    */
var RequirementQueryTypeEnum;
(function (RequirementQueryTypeEnum) {
    RequirementQueryTypeEnum["EPIC"] = "epic";
    RequirementQueryTypeEnum["USER_STORY"] = "user-story";
    RequirementQueryTypeEnum["FEATURE"] = "feature";
})(RequirementQueryTypeEnum || (exports.RequirementQueryTypeEnum = RequirementQueryTypeEnum = {}));
