// qase-api-client/dist/model/requirement.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Qase.io TestOps API v1
 * Qase TestOps API v1 Specification.
 *
 * The version of the OpenAPI document: 1.0.0
 * Contact: support@qase.io
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.RequirementTypeEnum = exports.RequirementStatusEnum = void 0;
/**
    * @export
    * @enum {string}
    */
var RequirementStatusEnum;
(function (RequirementStatusEnum) {
    RequirementStatusEnum["VALID"] = "valid";
    RequirementStatusEnum["DRAFT"] = "draft";
    RequirementStatusEnum["REVIEW"] = "review";
    RequirementStatusEnum["REWORK"] = "rework";
    RequirementStatusEnum["FINISH"] = "finish";
    RequirementStatusEnum["IMPLEMENTED"] = "implemented";
    RequirementStatusEnum["NOT_TESTABLE"] = "not-testable";
    RequirementStatusEnum["OBSOLETE"] = "obsolete";
})(RequirementStatusEnum || (exports.RequirementStatusEnum = RequirementStatusEnum = {}));
/**
    * @export
    * @enum {string}
    */
var RequirementTypeEnum;
(function (RequirementTypeEnum) {
    RequirementTypeEnum["EPIC"] = "epic";
    RequirementTypeEnum["USER_STORY"] = "user-story";
    RequirementTypeEnum["FEATURE"] = "feature";
})(RequirementTypeEnum || (exports.RequirementTypeEnum = RequirementTypeEnum = {}));
