// qase-api-client/dist/api/custom-fields-api.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Qase.io TestOps API v1
 * Qase TestOps API v1 Specification.
 *
 * The version of the OpenAPI document: 1.0.0
 * Contact: support@qase.io
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.CustomFieldsApi = exports.CustomFieldsApiFactory = exports.CustomFieldsApiFp = exports.CustomFieldsApiAxiosParamCreator = void 0;
const axios_1 = require("axios");
// Some imports not used depending on template conditions
// @ts-ignore
const common_1 = require("../common");
// @ts-ignore
const base_1 = require("../base");
/**
 * CustomFieldsApi - axios parameter creator
 * @export
 */
const CustomFieldsApiAxiosParamCreator = function (configuration) {
    return {
        /**
         * This method allows to create custom field.
         * @summary Create new Custom Field
         * @param {CustomFieldCreate} customFieldCreate
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        createCustomField: (customFieldCreate_1, ...args_1) => __awaiter(this, [customFieldCreate_1, ...args_1], void 0, function* (customFieldCreate, options = {}) {
            // verify required parameter 'customFieldCreate' is not null or undefined
            (0, common_1.assertParamExists)('createCustomField', 'customFieldCreate', customFieldCreate);
            const localVarPath = `/custom_field`;
            // use dummy base URL string because the URL constructor only accepts absolute URLs.
            const localVarUrlObj = new URL(localVarPath, common_1.DUMMY_BASE_URL);
            let baseOptions;
            if (configuration) {
                baseOptions = configuration.baseOptions;
            }
            const localVarRequestOptions = Object.assign(Object.assign({ method: 'POST' }, baseOptions), options);
            const localVarHeaderParameter = {};
            const localVarQueryParameter = {};
            // authentication TokenAuth required
            yield (0, common_1.setApiKeyToObject)(localVarHeaderParameter, "Token", configuration);
            localVarHeaderParameter['Content-Type'] = 'application/json';
            (0, common_1.setSearchParams)(localVarUrlObj, localVarQueryParameter);
            let headersFromBaseOptions = baseOptions && baseOptions.headers ? baseOptions.headers : {};
            localVarRequestOptions.headers = Object.assign(Object.assign(Object.assign({}, localVarHeaderParameter), headersFromBaseOptions), options.headers);
            localVarRequestOptions.data = (0, common_1.serializeDataIfNeeded)(customFieldCreate, localVarRequestOptions, configuration);
            return {
                url: (0, common_1.toPathString)(localVarUrlObj),
                options: localVarRequestOptions,
            };
        }),
        /**
         * This method allows to delete custom field.
         * @summary Delete Custom Field
         * @param {number} id Identifier.
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        deleteCustomField: (id_1, ...args_1) => __awaiter(this, [id_1, ...args_1], void 0, function* (id, options = {}) {
            // verify required parameter 'id' is not null or undefined
            (0, common_1.assertParamExists)('deleteCustomField', 'id', id);
            const localVarPath = `/custom_field/{id}`
                .replace(`{${"id"}}`, encodeURIComponent(String(id)));
            // use dummy base URL string because the URL constructor only accepts absolute URLs.
            const localVarUrlObj = new URL(localVarPath, common_1.DUMMY_BASE_URL);
            let baseOptions;
            if (configuration) {
                baseOptions = configuration.baseOptions;
            }
            const localVarRequestOptions = Object.assign(Object.assign({ method: 'DELETE' }, baseOptions), options);
            const localVarHeaderParameter = {};
            const localVarQueryParameter = {};
            // authentication TokenAuth required
            yield (0, common_1.setApiKeyToObject)(localVarHeaderParameter, "Token", configuration);
            (0, common_1.setSearchParams)(localVarUrlObj, localVarQueryParameter);
            let headersFromBaseOptions = baseOptions && baseOptions.headers ? baseOptions.headers : {};
            localVarRequestOptions.headers = Object.assign(Object.assign(Object.assign({}, localVarHeaderParameter), headersFromBaseOptions), options.headers);
            return {
                url: (0, common_1.toPathString)(localVarUrlObj),
                options: localVarRequestOptions,
            };
        }),
        /**
         * This method allows to retrieve custom field.
         * @summary Get Custom Field
         * @param {number} id Identifier.
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        getCustomField: (id_1, ...args_1) => __awaiter(this, [id_1, ...args_1], void 0, function* (id, options = {}) {
            // verify required parameter 'id' is not null or undefined
            (0, common_1.assertParamExists)('getCustomField', 'id', id);
            const localVarPath = `/custom_field/{id}`
                .replace(`{${"id"}}`, encodeURIComponent(String(id)));
            // use dummy base URL string because the URL constructor only accepts absolute URLs.
            const localVarUrlObj = new URL(localVarPath, common_1.DUMMY_BASE_URL);
            let baseOptions;
            if (configuration) {
                baseOptions = configuration.baseOptions;
            }
            const localVarRequestOptions = Object.assign(Object.assign({ method: 'GET' }, baseOptions), options);
            const localVarHeaderParameter = {};
            const localVarQueryParameter = {};
            // authentication TokenAuth required
            yield (0, common_1.setApiKeyToObject)(localVarHeaderParameter, "Token", configuration);
            (0, common_1.setSearchParams)(localVarUrlObj, localVarQueryParameter);
            let headersFromBaseOptions = baseOptions && baseOptions.headers ? baseOptions.headers : {};
            localVarRequestOptions.headers = Object.assign(Object.assign(Object.assign({}, localVarHeaderParameter), headersFromBaseOptions), options.headers);
            return {
                url: (0, common_1.toPathString)(localVarUrlObj),
                options: localVarRequestOptions,
            };
        }),
        /**
         * This method allows to retrieve and filter custom fields.
         * @summary Get all Custom Fields
         * @param {'case' | 'run' | 'defect'} [entity]
         * @param {'string' | 'text' | 'number' | 'checkbox' | 'selectbox' | 'radio' | 'multiselect' | 'url' | 'user' | 'datetime'} [type]
         * @param {number} [limit] A number of entities in result set.
         * @param {number} [offset] How many entities should be skipped.
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        getCustomFields: (entity_1, type_1, limit_1, offset_1, ...args_1) => __awaiter(this, [entity_1, type_1, limit_1, offset_1, ...args_1], void 0, function* (entity, type, limit, offset, options = {}) {
            const localVarPath = `/custom_field`;
            // use dummy base URL string because the URL constructor only accepts absolute URLs.
            const localVarUrlObj = new URL(localVarPath, common_1.DUMMY_BASE_URL);
            let baseOptions;
            if (configuration) {
                baseOptions = configuration.baseOptions;
            }
            const localVarRequestOptions = Object.assign(Object.assign({ method: 'GET' }, baseOptions), options);
            const localVarHeaderParameter = {};
            const localVarQueryParameter = {};
            // authentication TokenAuth required
            yield (0, common_1.setApiKeyToObject)(localVarHeaderParameter, "Token", configuration);
            if (entity !== undefined) {
                localVarQueryParameter['entity'] = entity;
            }
            if (type !== undefined) {
                localVarQueryParameter['type'] = type;
            }
            if (limit !== undefined) {
                localVarQueryParameter['limit'] = limit;
            }
            if (offset !== undefined) {
                localVarQueryParameter['offset'] = offset;
            }
            (0, common_1.setSearchParams)(localVarUrlObj, localVarQueryParameter);
            let headersFromBaseOptions = baseOptions && baseOptions.headers ? baseOptions.headers : {};
            localVarRequestOptions.headers = Object.assign(Object.assign(Object.assign({}, localVarHeaderParameter), headersFromBaseOptions), options.headers);
            return {
                url: (0, common_1.toPathString)(localVarUrlObj),
                options: localVarRequestOptions,
            };
        }),
        /**
         * This method allows to update custom field.
         * @summary Update Custom Field
         * @param {number} id Identifier.
         * @param {CustomFieldUpdate} customFieldUpdate
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        updateCustomField: (id_1, customFieldUpdate_1, ...args_1) => __awaiter(this, [id_1, customFieldUpdate_1, ...args_1], void 0, function* (id, customFieldUpdate, options = {}) {
            // verify required parameter 'id' is not null or undefined
            (0, common_1.assertParamExists)('updateCustomField', 'id', id);
            // verify required parameter 'customFieldUpdate' is not null or undefined
            (0, common_1.assertParamExists)('updateCustomField', 'customFieldUpdate', customFieldUpdate);
            const localVarPath = `/custom_field/{id}`
                .replace(`{${"id"}}`, encodeURIComponent(String(id)));
            // use dummy base URL string because the URL constructor only accepts absolute URLs.
            const localVarUrlObj = new URL(localVarPath, common_1.DUMMY_BASE_URL);
            let baseOptions;
            if (configuration) {
                baseOptions = configuration.baseOptions;
            }
            const localVarRequestOptions = Object.assign(Object.assign({ method: 'PATCH' }, baseOptions), options);
            const localVarHeaderParameter = {};
            const localVarQueryParameter = {};
            // authentication TokenAuth required
            yield (0, common_1.setApiKeyToObject)(localVarHeaderParameter, "Token", configuration);
            localVarHeaderParameter['Content-Type'] = 'application/json';
            (0, common_1.setSearchParams)(localVarUrlObj, localVarQueryParameter);
            let headersFromBaseOptions = baseOptions && baseOptions.headers ? baseOptions.headers : {};
            localVarRequestOptions.headers = Object.assign(Object.assign(Object.assign({}, localVarHeaderParameter), headersFromBaseOptions), options.headers);
            localVarRequestOptions.data = (0, common_1.serializeDataIfNeeded)(customFieldUpdate, localVarRequestOptions, configuration);
            return {
                url: (0, common_1.toPathString)(localVarUrlObj),
                options: localVarRequestOptions,
            };
        }),
    };
};
exports.CustomFieldsApiAxiosParamCreator = CustomFieldsApiAxiosParamCreator;
/**
 * CustomFieldsApi - functional programming interface
 * @export
 */
const CustomFieldsApiFp = function (configuration) {
    const localVarAxiosParamCreator = (0, exports.CustomFieldsApiAxiosParamCreator)(configuration);
    return {
        /**
         * This method allows to create custom field.
         * @summary Create new Custom Field
         * @param {CustomFieldCreate} customFieldCreate
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        createCustomField(customFieldCreate, options) {
            return __awaiter(this, void 0, void 0, function* () {
                const localVarAxiosArgs = yield localVarAxiosParamCreator.createCustomField(customFieldCreate, options);
                return (0, common_1.createRequestFunction)(localVarAxiosArgs, axios_1.default, base_1.BASE_PATH, configuration);
            });
        },
        /**
         * This method allows to delete custom field.
         * @summary Delete Custom Field
         * @param {number} id Identifier.
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        deleteCustomField(id, options) {
            return __awaiter(this, void 0, void 0, function* () {
                const localVarAxiosArgs = yield localVarAxiosParamCreator.deleteCustomField(id, options);
                return (0, common_1.createRequestFunction)(localVarAxiosArgs, axios_1.default, base_1.BASE_PATH, configuration);
            });
        },
        /**
         * This method allows to retrieve custom field.
         * @summary Get Custom Field
         * @param {number} id Identifier.
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        getCustomField(id, options) {
            return __awaiter(this, void 0, void 0, function* () {
                const localVarAxiosArgs = yield localVarAxiosParamCreator.getCustomField(id, options);
                return (0, common_1.createRequestFunction)(localVarAxiosArgs, axios_1.default, base_1.BASE_PATH, configuration);
            });
        },
        /**
         * This method allows to retrieve and filter custom fields.
         * @summary Get all Custom Fields
         * @param {'case' | 'run' | 'defect'} [entity]
         * @param {'string' | 'text' | 'number' | 'checkbox' | 'selectbox' | 'radio' | 'multiselect' | 'url' | 'user' | 'datetime'} [type]
         * @param {number} [limit] A number of entities in result set.
         * @param {number} [offset] How many entities should be skipped.
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        getCustomFields(entity, type, limit, offset, options) {
            return __awaiter(this, void 0, void 0, function* () {
                const localVarAxiosArgs = yield localVarAxiosParamCreator.getCustomFields(entity, type, limit, offset, options);
                return (0, common_1.createRequestFunction)(localVarAxiosArgs, axios_1.default, base_1.BASE_PATH, configuration);
            });
        },
        /**
         * This method allows to update custom field.
         * @summary Update Custom Field
         * @param {number} id Identifier.
         * @param {CustomFieldUpdate} customFieldUpdate
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        updateCustomField(id, customFieldUpdate, options) {
            return __awaiter(this, void 0, void 0, function* () {
                const localVarAxiosArgs = yield localVarAxiosParamCreator.updateCustomField(id, customFieldUpdate, options);
                return (0, common_1.createRequestFunction)(localVarAxiosArgs, axios_1.default, base_1.BASE_PATH, configuration);
            });
        },
    };
};
exports.CustomFieldsApiFp = CustomFieldsApiFp;
/**
 * CustomFieldsApi - factory interface
 * @export
 */
const CustomFieldsApiFactory = function (configuration, basePath, axios) {
    const localVarFp = (0, exports.CustomFieldsApiFp)(configuration);
    return {
        /**
         * This method allows to create custom field.
         * @summary Create new Custom Field
         * @param {CustomFieldCreate} customFieldCreate
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        createCustomField(customFieldCreate, options) {
            return localVarFp.createCustomField(customFieldCreate, options).then((request) => request(axios, basePath));
        },
        /**
         * This method allows to delete custom field.
         * @summary Delete Custom Field
         * @param {number} id Identifier.
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        deleteCustomField(id, options) {
            return localVarFp.deleteCustomField(id, options).then((request) => request(axios, basePath));
        },
        /**
         * This method allows to retrieve custom field.
         * @summary Get Custom Field
         * @param {number} id Identifier.
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        getCustomField(id, options) {
            return localVarFp.getCustomField(id, options).then((request) => request(axios, basePath));
        },
        /**
         * This method allows to retrieve and filter custom fields.
         * @summary Get all Custom Fields
         * @param {'case' | 'run' | 'defect'} [entity]
         * @param {'string' | 'text' | 'number' | 'checkbox' | 'selectbox' | 'radio' | 'multiselect' | 'url' | 'user' | 'datetime'} [type]
         * @param {number} [limit] A number of entities in result set.
         * @param {number} [offset] How many entities should be skipped.
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        getCustomFields(entity, type, limit, offset, options) {
            return localVarFp.getCustomFields(entity, type, limit, offset, options).then((request) => request(axios, basePath));
        },
        /**
         * This method allows to update custom field.
         * @summary Update Custom Field
         * @param {number} id Identifier.
         * @param {CustomFieldUpdate} customFieldUpdate
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        updateCustomField(id, customFieldUpdate, options) {
            return localVarFp.updateCustomField(id, customFieldUpdate, options).then((request) => request(axios, basePath));
        },
    };
};
exports.CustomFieldsApiFactory = CustomFieldsApiFactory;
/**
 * CustomFieldsApi - object-oriented interface
 * @export
 * @class CustomFieldsApi
 * @extends {BaseAPI}
 */
class CustomFieldsApi extends base_1.BaseAPI {
    /**
     * This method allows to create custom field.
     * @summary Create new Custom Field
     * @param {CustomFieldCreate} customFieldCreate
     * @param {*} [options] Override http request option.
     * @throws {RequiredError}
     * @memberof CustomFieldsApi
     */
    createCustomField(customFieldCreate, options) {
        return (0, exports.CustomFieldsApiFp)(this.configuration).createCustomField(customFieldCreate, options).then((request) => request(this.axios, this.basePath));
    }
    /**
     * This method allows to delete custom field.
     * @summary Delete Custom Field
     * @param {number} id Identifier.
     * @param {*} [options] Override http request option.
     * @throws {RequiredError}
     * @memberof CustomFieldsApi
     */
    deleteCustomField(id, options) {
        return (0, exports.CustomFieldsApiFp)(this.configuration).deleteCustomField(id, options).then((request) => request(this.axios, this.basePath));
    }
    /**
     * This method allows to retrieve custom field.
     * @summary Get Custom Field
     * @param {number} id Identifier.
     * @param {*} [options] Override http request option.
     * @throws {RequiredError}
     * @memberof CustomFieldsApi
     */
    getCustomField(id, options) {
        return (0, exports.CustomFieldsApiFp)(this.configuration).getCustomField(id, options).then((request) => request(this.axios, this.basePath));
    }
    /**
     * This method allows to retrieve and filter custom fields.
     * @summary Get all Custom Fields
     * @param {'case' | 'run' | 'defect'} [entity]
     * @param {'string' | 'text' | 'number' | 'checkbox' | 'selectbox' | 'radio' | 'multiselect' | 'url' | 'user' | 'datetime'} [type]
     * @param {number} [limit] A number of entities in result set.
     * @param {number} [offset] How many entities should be skipped.
     * @param {*} [options] Override http request option.
     * @throws {RequiredError}
     * @memberof CustomFieldsApi
     */
    getCustomFields(entity, type, limit, offset, options) {
        return (0, exports.CustomFieldsApiFp)(this.configuration).getCustomFields(entity, type, limit, offset, options).then((request) => request(this.axios, this.basePath));
    }
    /**
     * This method allows to update custom field.
     * @summary Update Custom Field
     * @param {number} id Identifier.
     * @param {CustomFieldUpdate} customFieldUpdate
     * @param {*} [options] Override http request option.
     * @throws {RequiredError}
     * @memberof CustomFieldsApi
     */
    updateCustomField(id, customFieldUpdate, options) {
        return (0, exports.CustomFieldsApiFp)(this.configuration).updateCustomField(id, customFieldUpdate, options).then((request) => request(this.axios, this.basePath));
    }
}
exports.CustomFieldsApi = CustomFieldsApi;
