// qase-api-client/dist/api/environments-api.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Qase.io TestOps API v1
 * Qase TestOps API v1 Specification.
 *
 * The version of the OpenAPI document: 1.0.0
 * Contact: support@qase.io
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.EnvironmentsApi = exports.EnvironmentsApiFactory = exports.EnvironmentsApiFp = exports.EnvironmentsApiAxiosParamCreator = void 0;
const axios_1 = require("axios");
// Some imports not used depending on template conditions
// @ts-ignore
const common_1 = require("../common");
// @ts-ignore
const base_1 = require("../base");
/**
 * EnvironmentsApi - axios parameter creator
 * @export
 */
const EnvironmentsApiAxiosParamCreator = function (configuration) {
    return {
        /**
         * This method allows to create an environment in selected project.
         * @summary Create a new environment
         * @param {string} code Code of project, where to search entities.
         * @param {EnvironmentCreate} environmentCreate
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        createEnvironment: (code_1, environmentCreate_1, ...args_1) => __awaiter(this, [code_1, environmentCreate_1, ...args_1], void 0, function* (code, environmentCreate, options = {}) {
            // verify required parameter 'code' is not null or undefined
            (0, common_1.assertParamExists)('createEnvironment', 'code', code);
            // verify required parameter 'environmentCreate' is not null or undefined
            (0, common_1.assertParamExists)('createEnvironment', 'environmentCreate', environmentCreate);
            const localVarPath = `/environment/{code}`
                .replace(`{${"code"}}`, encodeURIComponent(String(code)));
            // use dummy base URL string because the URL constructor only accepts absolute URLs.
            const localVarUrlObj = new URL(localVarPath, common_1.DUMMY_BASE_URL);
            let baseOptions;
            if (configuration) {
                baseOptions = configuration.baseOptions;
            }
            const localVarRequestOptions = Object.assign(Object.assign({ method: 'POST' }, baseOptions), options);
            const localVarHeaderParameter = {};
            const localVarQueryParameter = {};
            // authentication TokenAuth required
            yield (0, common_1.setApiKeyToObject)(localVarHeaderParameter, "Token", configuration);
            localVarHeaderParameter['Content-Type'] = 'application/json';
            (0, common_1.setSearchParams)(localVarUrlObj, localVarQueryParameter);
            let headersFromBaseOptions = baseOptions && baseOptions.headers ? baseOptions.headers : {};
            localVarRequestOptions.headers = Object.assign(Object.assign(Object.assign({}, localVarHeaderParameter), headersFromBaseOptions), options.headers);
            localVarRequestOptions.data = (0, common_1.serializeDataIfNeeded)(environmentCreate, localVarRequestOptions, configuration);
            return {
                url: (0, common_1.toPathString)(localVarUrlObj),
                options: localVarRequestOptions,
            };
        }),
        /**
         * This method completely deletes an environment from repository.
         * @summary Delete environment
         * @param {string} code Code of project, where to search entities.
         * @param {number} id Identifier.
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        deleteEnvironment: (code_1, id_1, ...args_1) => __awaiter(this, [code_1, id_1, ...args_1], void 0, function* (code, id, options = {}) {
            // verify required parameter 'code' is not null or undefined
            (0, common_1.assertParamExists)('deleteEnvironment', 'code', code);
            // verify required parameter 'id' is not null or undefined
            (0, common_1.assertParamExists)('deleteEnvironment', 'id', id);
            const localVarPath = `/environment/{code}/{id}`
                .replace(`{${"code"}}`, encodeURIComponent(String(code)))
                .replace(`{${"id"}}`, encodeURIComponent(String(id)));
            // use dummy base URL string because the URL constructor only accepts absolute URLs.
            const localVarUrlObj = new URL(localVarPath, common_1.DUMMY_BASE_URL);
            let baseOptions;
            if (configuration) {
                baseOptions = configuration.baseOptions;
            }
            const localVarRequestOptions = Object.assign(Object.assign({ method: 'DELETE' }, baseOptions), options);
            const localVarHeaderParameter = {};
            const localVarQueryParameter = {};
            // authentication TokenAuth required
            yield (0, common_1.setApiKeyToObject)(localVarHeaderParameter, "Token", configuration);
            (0, common_1.setSearchParams)(localVarUrlObj, localVarQueryParameter);
            let headersFromBaseOptions = baseOptions && baseOptions.headers ? baseOptions.headers : {};
            localVarRequestOptions.headers = Object.assign(Object.assign(Object.assign({}, localVarHeaderParameter), headersFromBaseOptions), options.headers);
            return {
                url: (0, common_1.toPathString)(localVarUrlObj),
                options: localVarRequestOptions,
            };
        }),
        /**
         * This method allows to retrieve a specific environment.
         * @summary Get a specific environment
         * @param {string} code Code of project, where to search entities.
         * @param {number} id Identifier.
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        getEnvironment: (code_1, id_1, ...args_1) => __awaiter(this, [code_1, id_1, ...args_1], void 0, function* (code, id, options = {}) {
            // verify required parameter 'code' is not null or undefined
            (0, common_1.assertParamExists)('getEnvironment', 'code', code);
            // verify required parameter 'id' is not null or undefined
            (0, common_1.assertParamExists)('getEnvironment', 'id', id);
            const localVarPath = `/environment/{code}/{id}`
                .replace(`{${"code"}}`, encodeURIComponent(String(code)))
                .replace(`{${"id"}}`, encodeURIComponent(String(id)));
            // use dummy base URL string because the URL constructor only accepts absolute URLs.
            const localVarUrlObj = new URL(localVarPath, common_1.DUMMY_BASE_URL);
            let baseOptions;
            if (configuration) {
                baseOptions = configuration.baseOptions;
            }
            const localVarRequestOptions = Object.assign(Object.assign({ method: 'GET' }, baseOptions), options);
            const localVarHeaderParameter = {};
            const localVarQueryParameter = {};
            // authentication TokenAuth required
            yield (0, common_1.setApiKeyToObject)(localVarHeaderParameter, "Token", configuration);
            (0, common_1.setSearchParams)(localVarUrlObj, localVarQueryParameter);
            let headersFromBaseOptions = baseOptions && baseOptions.headers ? baseOptions.headers : {};
            localVarRequestOptions.headers = Object.assign(Object.assign(Object.assign({}, localVarHeaderParameter), headersFromBaseOptions), options.headers);
            return {
                url: (0, common_1.toPathString)(localVarUrlObj),
                options: localVarRequestOptions,
            };
        }),
        /**
         * This method allows to retrieve all environments stored in selected project.
         * @summary Get all environments
         * @param {string} code Code of project, where to search entities.
         * @param {string} [search] A search string. Will return all environments with titles containing provided string.
         * @param {string} [slug] A search string.  Will return all environments with slugs equal to provided string.
         * @param {number} [limit] A number of entities in result set.
         * @param {number} [offset] How many entities should be skipped.
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        getEnvironments: (code_1, search_1, slug_1, limit_1, offset_1, ...args_1) => __awaiter(this, [code_1, search_1, slug_1, limit_1, offset_1, ...args_1], void 0, function* (code, search, slug, limit, offset, options = {}) {
            // verify required parameter 'code' is not null or undefined
            (0, common_1.assertParamExists)('getEnvironments', 'code', code);
            const localVarPath = `/environment/{code}`
                .replace(`{${"code"}}`, encodeURIComponent(String(code)));
            // use dummy base URL string because the URL constructor only accepts absolute URLs.
            const localVarUrlObj = new URL(localVarPath, common_1.DUMMY_BASE_URL);
            let baseOptions;
            if (configuration) {
                baseOptions = configuration.baseOptions;
            }
            const localVarRequestOptions = Object.assign(Object.assign({ method: 'GET' }, baseOptions), options);
            const localVarHeaderParameter = {};
            const localVarQueryParameter = {};
            // authentication TokenAuth required
            yield (0, common_1.setApiKeyToObject)(localVarHeaderParameter, "Token", configuration);
            if (search !== undefined) {
                localVarQueryParameter['search'] = search;
            }
            if (slug !== undefined) {
                localVarQueryParameter['slug'] = slug;
            }
            if (limit !== undefined) {
                localVarQueryParameter['limit'] = limit;
            }
            if (offset !== undefined) {
                localVarQueryParameter['offset'] = offset;
            }
            (0, common_1.setSearchParams)(localVarUrlObj, localVarQueryParameter);
            let headersFromBaseOptions = baseOptions && baseOptions.headers ? baseOptions.headers : {};
            localVarRequestOptions.headers = Object.assign(Object.assign(Object.assign({}, localVarHeaderParameter), headersFromBaseOptions), options.headers);
            return {
                url: (0, common_1.toPathString)(localVarUrlObj),
                options: localVarRequestOptions,
            };
        }),
        /**
         * This method updates an environment.
         * @summary Update environment
         * @param {string} code Code of project, where to search entities.
         * @param {number} id Identifier.
         * @param {EnvironmentUpdate} environmentUpdate
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        updateEnvironment: (code_1, id_1, environmentUpdate_1, ...args_1) => __awaiter(this, [code_1, id_1, environmentUpdate_1, ...args_1], void 0, function* (code, id, environmentUpdate, options = {}) {
            // verify required parameter 'code' is not null or undefined
            (0, common_1.assertParamExists)('updateEnvironment', 'code', code);
            // verify required parameter 'id' is not null or undefined
            (0, common_1.assertParamExists)('updateEnvironment', 'id', id);
            // verify required parameter 'environmentUpdate' is not null or undefined
            (0, common_1.assertParamExists)('updateEnvironment', 'environmentUpdate', environmentUpdate);
            const localVarPath = `/environment/{code}/{id}`
                .replace(`{${"code"}}`, encodeURIComponent(String(code)))
                .replace(`{${"id"}}`, encodeURIComponent(String(id)));
            // use dummy base URL string because the URL constructor only accepts absolute URLs.
            const localVarUrlObj = new URL(localVarPath, common_1.DUMMY_BASE_URL);
            let baseOptions;
            if (configuration) {
                baseOptions = configuration.baseOptions;
            }
            const localVarRequestOptions = Object.assign(Object.assign({ method: 'PATCH' }, baseOptions), options);
            const localVarHeaderParameter = {};
            const localVarQueryParameter = {};
            // authentication TokenAuth required
            yield (0, common_1.setApiKeyToObject)(localVarHeaderParameter, "Token", configuration);
            localVarHeaderParameter['Content-Type'] = 'application/json';
            (0, common_1.setSearchParams)(localVarUrlObj, localVarQueryParameter);
            let headersFromBaseOptions = baseOptions && baseOptions.headers ? baseOptions.headers : {};
            localVarRequestOptions.headers = Object.assign(Object.assign(Object.assign({}, localVarHeaderParameter), headersFromBaseOptions), options.headers);
            localVarRequestOptions.data = (0, common_1.serializeDataIfNeeded)(environmentUpdate, localVarRequestOptions, configuration);
            return {
                url: (0, common_1.toPathString)(localVarUrlObj),
                options: localVarRequestOptions,
            };
        }),
    };
};
exports.EnvironmentsApiAxiosParamCreator = EnvironmentsApiAxiosParamCreator;
/**
 * EnvironmentsApi - functional programming interface
 * @export
 */
const EnvironmentsApiFp = function (configuration) {
    const localVarAxiosParamCreator = (0, exports.EnvironmentsApiAxiosParamCreator)(configuration);
    return {
        /**
         * This method allows to create an environment in selected project.
         * @summary Create a new environment
         * @param {string} code Code of project, where to search entities.
         * @param {EnvironmentCreate} environmentCreate
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        createEnvironment(code, environmentCreate, options) {
            return __awaiter(this, void 0, void 0, function* () {
                const localVarAxiosArgs = yield localVarAxiosParamCreator.createEnvironment(code, environmentCreate, options);
                return (0, common_1.createRequestFunction)(localVarAxiosArgs, axios_1.default, base_1.BASE_PATH, configuration);
            });
        },
        /**
         * This method completely deletes an environment from repository.
         * @summary Delete environment
         * @param {string} code Code of project, where to search entities.
         * @param {number} id Identifier.
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        deleteEnvironment(code, id, options) {
            return __awaiter(this, void 0, void 0, function* () {
                const localVarAxiosArgs = yield localVarAxiosParamCreator.deleteEnvironment(code, id, options);
                return (0, common_1.createRequestFunction)(localVarAxiosArgs, axios_1.default, base_1.BASE_PATH, configuration);
            });
        },
        /**
         * This method allows to retrieve a specific environment.
         * @summary Get a specific environment
         * @param {string} code Code of project, where to search entities.
         * @param {number} id Identifier.
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        getEnvironment(code, id, options) {
            return __awaiter(this, void 0, void 0, function* () {
                const localVarAxiosArgs = yield localVarAxiosParamCreator.getEnvironment(code, id, options);
                return (0, common_1.createRequestFunction)(localVarAxiosArgs, axios_1.default, base_1.BASE_PATH, configuration);
            });
        },
        /**
         * This method allows to retrieve all environments stored in selected project.
         * @summary Get all environments
         * @param {string} code Code of project, where to search entities.
         * @param {string} [search] A search string. Will return all environments with titles containing provided string.
         * @param {string} [slug] A search string.  Will return all environments with slugs equal to provided string.
         * @param {number} [limit] A number of entities in result set.
         * @param {number} [offset] How many entities should be skipped.
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        getEnvironments(code, search, slug, limit, offset, options) {
            return __awaiter(this, void 0, void 0, function* () {
                const localVarAxiosArgs = yield localVarAxiosParamCreator.getEnvironments(code, search, slug, limit, offset, options);
                return (0, common_1.createRequestFunction)(localVarAxiosArgs, axios_1.default, base_1.BASE_PATH, configuration);
            });
        },
        /**
         * This method updates an environment.
         * @summary Update environment
         * @param {string} code Code of project, where to search entities.
         * @param {number} id Identifier.
         * @param {EnvironmentUpdate} environmentUpdate
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        updateEnvironment(code, id, environmentUpdate, options) {
            return __awaiter(this, void 0, void 0, function* () {
                const localVarAxiosArgs = yield localVarAxiosParamCreator.updateEnvironment(code, id, environmentUpdate, options);
                return (0, common_1.createRequestFunction)(localVarAxiosArgs, axios_1.default, base_1.BASE_PATH, configuration);
            });
        },
    };
};
exports.EnvironmentsApiFp = EnvironmentsApiFp;
/**
 * EnvironmentsApi - factory interface
 * @export
 */
const EnvironmentsApiFactory = function (configuration, basePath, axios) {
    const localVarFp = (0, exports.EnvironmentsApiFp)(configuration);
    return {
        /**
         * This method allows to create an environment in selected project.
         * @summary Create a new environment
         * @param {string} code Code of project, where to search entities.
         * @param {EnvironmentCreate} environmentCreate
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        createEnvironment(code, environmentCreate, options) {
            return localVarFp.createEnvironment(code, environmentCreate, options).then((request) => request(axios, basePath));
        },
        /**
         * This method completely deletes an environment from repository.
         * @summary Delete environment
         * @param {string} code Code of project, where to search entities.
         * @param {number} id Identifier.
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        deleteEnvironment(code, id, options) {
            return localVarFp.deleteEnvironment(code, id, options).then((request) => request(axios, basePath));
        },
        /**
         * This method allows to retrieve a specific environment.
         * @summary Get a specific environment
         * @param {string} code Code of project, where to search entities.
         * @param {number} id Identifier.
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        getEnvironment(code, id, options) {
            return localVarFp.getEnvironment(code, id, options).then((request) => request(axios, basePath));
        },
        /**
         * This method allows to retrieve all environments stored in selected project.
         * @summary Get all environments
         * @param {string} code Code of project, where to search entities.
         * @param {string} [search] A search string. Will return all environments with titles containing provided string.
         * @param {string} [slug] A search string.  Will return all environments with slugs equal to provided string.
         * @param {number} [limit] A number of entities in result set.
         * @param {number} [offset] How many entities should be skipped.
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        getEnvironments(code, search, slug, limit, offset, options) {
            return localVarFp.getEnvironments(code, search, slug, limit, offset, options).then((request) => request(axios, basePath));
        },
        /**
         * This method updates an environment.
         * @summary Update environment
         * @param {string} code Code of project, where to search entities.
         * @param {number} id Identifier.
         * @param {EnvironmentUpdate} environmentUpdate
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        updateEnvironment(code, id, environmentUpdate, options) {
            return localVarFp.updateEnvironment(code, id, environmentUpdate, options).then((request) => request(axios, basePath));
        },
    };
};
exports.EnvironmentsApiFactory = EnvironmentsApiFactory;
/**
 * EnvironmentsApi - object-oriented interface
 * @export
 * @class EnvironmentsApi
 * @extends {BaseAPI}
 */
class EnvironmentsApi extends base_1.BaseAPI {
    /**
     * This method allows to create an environment in selected project.
     * @summary Create a new environment
     * @param {string} code Code of project, where to search entities.
     * @param {EnvironmentCreate} environmentCreate
     * @param {*} [options] Override http request option.
     * @throws {RequiredError}
     * @memberof EnvironmentsApi
     */
    createEnvironment(code, environmentCreate, options) {
        return (0, exports.EnvironmentsApiFp)(this.configuration).createEnvironment(code, environmentCreate, options).then((request) => request(this.axios, this.basePath));
    }
    /**
     * This method completely deletes an environment from repository.
     * @summary Delete environment
     * @param {string} code Code of project, where to search entities.
     * @param {number} id Identifier.
     * @param {*} [options] Override http request option.
     * @throws {RequiredError}
     * @memberof EnvironmentsApi
     */
    deleteEnvironment(code, id, options) {
        return (0, exports.EnvironmentsApiFp)(this.configuration).deleteEnvironment(code, id, options).then((request) => request(this.axios, this.basePath));
    }
    /**
     * This method allows to retrieve a specific environment.
     * @summary Get a specific environment
     * @param {string} code Code of project, where to search entities.
     * @param {number} id Identifier.
     * @param {*} [options] Override http request option.
     * @throws {RequiredError}
     * @memberof EnvironmentsApi
     */
    getEnvironment(code, id, options) {
        return (0, exports.EnvironmentsApiFp)(this.configuration).getEnvironment(code, id, options).then((request) => request(this.axios, this.basePath));
    }
    /**
     * This method allows to retrieve all environments stored in selected project.
     * @summary Get all environments
     * @param {string} code Code of project, where to search entities.
     * @param {string} [search] A search string. Will return all environments with titles containing provided string.
     * @param {string} [slug] A search string.  Will return all environments with slugs equal to provided string.
     * @param {number} [limit] A number of entities in result set.
     * @param {number} [offset] How many entities should be skipped.
     * @param {*} [options] Override http request option.
     * @throws {RequiredError}
     * @memberof EnvironmentsApi
     */
    getEnvironments(code, search, slug, limit, offset, options) {
        return (0, exports.EnvironmentsApiFp)(this.configuration).getEnvironments(code, search, slug, limit, offset, options).then((request) => request(this.axios, this.basePath));
    }
    /**
     * This method updates an environment.
     * @summary Update environment
     * @param {string} code Code of project, where to search entities.
     * @param {number} id Identifier.
     * @param {EnvironmentUpdate} environmentUpdate
     * @param {*} [options] Override http request option.
     * @throws {RequiredError}
     * @memberof EnvironmentsApi
     */
    updateEnvironment(code, id, environmentUpdate, options) {
        return (0, exports.EnvironmentsApiFp)(this.configuration).updateEnvironment(code, id, environmentUpdate, options).then((request) => request(this.axios, this.basePath));
    }
}
exports.EnvironmentsApi = EnvironmentsApi;
