// qase-api-client/dist/api/plans-api.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Qase.io TestOps API v1
 * Qase TestOps API v1 Specification.
 *
 * The version of the OpenAPI document: 1.0.0
 * Contact: support@qase.io
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.PlansApi = exports.PlansApiFactory = exports.PlansApiFp = exports.PlansApiAxiosParamCreator = void 0;
const axios_1 = require("axios");
// Some imports not used depending on template conditions
// @ts-ignore
const common_1 = require("../common");
// @ts-ignore
const base_1 = require("../base");
/**
 * PlansApi - axios parameter creator
 * @export
 */
const PlansApiAxiosParamCreator = function (configuration) {
    return {
        /**
         * This method allows to create a plan in selected project.
         * @summary Create a new plan
         * @param {string} code Code of project, where to search entities.
         * @param {PlanCreate} planCreate
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        createPlan: (code_1, planCreate_1, ...args_1) => __awaiter(this, [code_1, planCreate_1, ...args_1], void 0, function* (code, planCreate, options = {}) {
            // verify required parameter 'code' is not null or undefined
            (0, common_1.assertParamExists)('createPlan', 'code', code);
            // verify required parameter 'planCreate' is not null or undefined
            (0, common_1.assertParamExists)('createPlan', 'planCreate', planCreate);
            const localVarPath = `/plan/{code}`
                .replace(`{${"code"}}`, encodeURIComponent(String(code)));
            // use dummy base URL string because the URL constructor only accepts absolute URLs.
            const localVarUrlObj = new URL(localVarPath, common_1.DUMMY_BASE_URL);
            let baseOptions;
            if (configuration) {
                baseOptions = configuration.baseOptions;
            }
            const localVarRequestOptions = Object.assign(Object.assign({ method: 'POST' }, baseOptions), options);
            const localVarHeaderParameter = {};
            const localVarQueryParameter = {};
            // authentication TokenAuth required
            yield (0, common_1.setApiKeyToObject)(localVarHeaderParameter, "Token", configuration);
            localVarHeaderParameter['Content-Type'] = 'application/json';
            (0, common_1.setSearchParams)(localVarUrlObj, localVarQueryParameter);
            let headersFromBaseOptions = baseOptions && baseOptions.headers ? baseOptions.headers : {};
            localVarRequestOptions.headers = Object.assign(Object.assign(Object.assign({}, localVarHeaderParameter), headersFromBaseOptions), options.headers);
            localVarRequestOptions.data = (0, common_1.serializeDataIfNeeded)(planCreate, localVarRequestOptions, configuration);
            return {
                url: (0, common_1.toPathString)(localVarUrlObj),
                options: localVarRequestOptions,
            };
        }),
        /**
         * This method completely deletes a plan from repository.
         * @summary Delete plan
         * @param {string} code Code of project, where to search entities.
         * @param {number} id Identifier.
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        deletePlan: (code_1, id_1, ...args_1) => __awaiter(this, [code_1, id_1, ...args_1], void 0, function* (code, id, options = {}) {
            // verify required parameter 'code' is not null or undefined
            (0, common_1.assertParamExists)('deletePlan', 'code', code);
            // verify required parameter 'id' is not null or undefined
            (0, common_1.assertParamExists)('deletePlan', 'id', id);
            const localVarPath = `/plan/{code}/{id}`
                .replace(`{${"code"}}`, encodeURIComponent(String(code)))
                .replace(`{${"id"}}`, encodeURIComponent(String(id)));
            // use dummy base URL string because the URL constructor only accepts absolute URLs.
            const localVarUrlObj = new URL(localVarPath, common_1.DUMMY_BASE_URL);
            let baseOptions;
            if (configuration) {
                baseOptions = configuration.baseOptions;
            }
            const localVarRequestOptions = Object.assign(Object.assign({ method: 'DELETE' }, baseOptions), options);
            const localVarHeaderParameter = {};
            const localVarQueryParameter = {};
            // authentication TokenAuth required
            yield (0, common_1.setApiKeyToObject)(localVarHeaderParameter, "Token", configuration);
            (0, common_1.setSearchParams)(localVarUrlObj, localVarQueryParameter);
            let headersFromBaseOptions = baseOptions && baseOptions.headers ? baseOptions.headers : {};
            localVarRequestOptions.headers = Object.assign(Object.assign(Object.assign({}, localVarHeaderParameter), headersFromBaseOptions), options.headers);
            return {
                url: (0, common_1.toPathString)(localVarUrlObj),
                options: localVarRequestOptions,
            };
        }),
        /**
         * This method allows to retrieve a specific plan.
         * @summary Get a specific plan
         * @param {string} code Code of project, where to search entities.
         * @param {number} id Identifier.
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        getPlan: (code_1, id_1, ...args_1) => __awaiter(this, [code_1, id_1, ...args_1], void 0, function* (code, id, options = {}) {
            // verify required parameter 'code' is not null or undefined
            (0, common_1.assertParamExists)('getPlan', 'code', code);
            // verify required parameter 'id' is not null or undefined
            (0, common_1.assertParamExists)('getPlan', 'id', id);
            const localVarPath = `/plan/{code}/{id}`
                .replace(`{${"code"}}`, encodeURIComponent(String(code)))
                .replace(`{${"id"}}`, encodeURIComponent(String(id)));
            // use dummy base URL string because the URL constructor only accepts absolute URLs.
            const localVarUrlObj = new URL(localVarPath, common_1.DUMMY_BASE_URL);
            let baseOptions;
            if (configuration) {
                baseOptions = configuration.baseOptions;
            }
            const localVarRequestOptions = Object.assign(Object.assign({ method: 'GET' }, baseOptions), options);
            const localVarHeaderParameter = {};
            const localVarQueryParameter = {};
            // authentication TokenAuth required
            yield (0, common_1.setApiKeyToObject)(localVarHeaderParameter, "Token", configuration);
            (0, common_1.setSearchParams)(localVarUrlObj, localVarQueryParameter);
            let headersFromBaseOptions = baseOptions && baseOptions.headers ? baseOptions.headers : {};
            localVarRequestOptions.headers = Object.assign(Object.assign(Object.assign({}, localVarHeaderParameter), headersFromBaseOptions), options.headers);
            return {
                url: (0, common_1.toPathString)(localVarUrlObj),
                options: localVarRequestOptions,
            };
        }),
        /**
         * This method allows to retrieve all plans stored in selected project.
         * @summary Get all plans
         * @param {string} code Code of project, where to search entities.
         * @param {number} [limit] A number of entities in result set.
         * @param {number} [offset] How many entities should be skipped.
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        getPlans: (code_1, limit_1, offset_1, ...args_1) => __awaiter(this, [code_1, limit_1, offset_1, ...args_1], void 0, function* (code, limit, offset, options = {}) {
            // verify required parameter 'code' is not null or undefined
            (0, common_1.assertParamExists)('getPlans', 'code', code);
            const localVarPath = `/plan/{code}`
                .replace(`{${"code"}}`, encodeURIComponent(String(code)));
            // use dummy base URL string because the URL constructor only accepts absolute URLs.
            const localVarUrlObj = new URL(localVarPath, common_1.DUMMY_BASE_URL);
            let baseOptions;
            if (configuration) {
                baseOptions = configuration.baseOptions;
            }
            const localVarRequestOptions = Object.assign(Object.assign({ method: 'GET' }, baseOptions), options);
            const localVarHeaderParameter = {};
            const localVarQueryParameter = {};
            // authentication TokenAuth required
            yield (0, common_1.setApiKeyToObject)(localVarHeaderParameter, "Token", configuration);
            if (limit !== undefined) {
                localVarQueryParameter['limit'] = limit;
            }
            if (offset !== undefined) {
                localVarQueryParameter['offset'] = offset;
            }
            (0, common_1.setSearchParams)(localVarUrlObj, localVarQueryParameter);
            let headersFromBaseOptions = baseOptions && baseOptions.headers ? baseOptions.headers : {};
            localVarRequestOptions.headers = Object.assign(Object.assign(Object.assign({}, localVarHeaderParameter), headersFromBaseOptions), options.headers);
            return {
                url: (0, common_1.toPathString)(localVarUrlObj),
                options: localVarRequestOptions,
            };
        }),
        /**
         * This method updates a plan.
         * @summary Update plan
         * @param {string} code Code of project, where to search entities.
         * @param {number} id Identifier.
         * @param {PlanUpdate} planUpdate
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        updatePlan: (code_1, id_1, planUpdate_1, ...args_1) => __awaiter(this, [code_1, id_1, planUpdate_1, ...args_1], void 0, function* (code, id, planUpdate, options = {}) {
            // verify required parameter 'code' is not null or undefined
            (0, common_1.assertParamExists)('updatePlan', 'code', code);
            // verify required parameter 'id' is not null or undefined
            (0, common_1.assertParamExists)('updatePlan', 'id', id);
            // verify required parameter 'planUpdate' is not null or undefined
            (0, common_1.assertParamExists)('updatePlan', 'planUpdate', planUpdate);
            const localVarPath = `/plan/{code}/{id}`
                .replace(`{${"code"}}`, encodeURIComponent(String(code)))
                .replace(`{${"id"}}`, encodeURIComponent(String(id)));
            // use dummy base URL string because the URL constructor only accepts absolute URLs.
            const localVarUrlObj = new URL(localVarPath, common_1.DUMMY_BASE_URL);
            let baseOptions;
            if (configuration) {
                baseOptions = configuration.baseOptions;
            }
            const localVarRequestOptions = Object.assign(Object.assign({ method: 'PATCH' }, baseOptions), options);
            const localVarHeaderParameter = {};
            const localVarQueryParameter = {};
            // authentication TokenAuth required
            yield (0, common_1.setApiKeyToObject)(localVarHeaderParameter, "Token", configuration);
            localVarHeaderParameter['Content-Type'] = 'application/json';
            (0, common_1.setSearchParams)(localVarUrlObj, localVarQueryParameter);
            let headersFromBaseOptions = baseOptions && baseOptions.headers ? baseOptions.headers : {};
            localVarRequestOptions.headers = Object.assign(Object.assign(Object.assign({}, localVarHeaderParameter), headersFromBaseOptions), options.headers);
            localVarRequestOptions.data = (0, common_1.serializeDataIfNeeded)(planUpdate, localVarRequestOptions, configuration);
            return {
                url: (0, common_1.toPathString)(localVarUrlObj),
                options: localVarRequestOptions,
            };
        }),
    };
};
exports.PlansApiAxiosParamCreator = PlansApiAxiosParamCreator;
/**
 * PlansApi - functional programming interface
 * @export
 */
const PlansApiFp = function (configuration) {
    const localVarAxiosParamCreator = (0, exports.PlansApiAxiosParamCreator)(configuration);
    return {
        /**
         * This method allows to create a plan in selected project.
         * @summary Create a new plan
         * @param {string} code Code of project, where to search entities.
         * @param {PlanCreate} planCreate
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        createPlan(code, planCreate, options) {
            return __awaiter(this, void 0, void 0, function* () {
                const localVarAxiosArgs = yield localVarAxiosParamCreator.createPlan(code, planCreate, options);
                return (0, common_1.createRequestFunction)(localVarAxiosArgs, axios_1.default, base_1.BASE_PATH, configuration);
            });
        },
        /**
         * This method completely deletes a plan from repository.
         * @summary Delete plan
         * @param {string} code Code of project, where to search entities.
         * @param {number} id Identifier.
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        deletePlan(code, id, options) {
            return __awaiter(this, void 0, void 0, function* () {
                const localVarAxiosArgs = yield localVarAxiosParamCreator.deletePlan(code, id, options);
                return (0, common_1.createRequestFunction)(localVarAxiosArgs, axios_1.default, base_1.BASE_PATH, configuration);
            });
        },
        /**
         * This method allows to retrieve a specific plan.
         * @summary Get a specific plan
         * @param {string} code Code of project, where to search entities.
         * @param {number} id Identifier.
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        getPlan(code, id, options) {
            return __awaiter(this, void 0, void 0, function* () {
                const localVarAxiosArgs = yield localVarAxiosParamCreator.getPlan(code, id, options);
                return (0, common_1.createRequestFunction)(localVarAxiosArgs, axios_1.default, base_1.BASE_PATH, configuration);
            });
        },
        /**
         * This method allows to retrieve all plans stored in selected project.
         * @summary Get all plans
         * @param {string} code Code of project, where to search entities.
         * @param {number} [limit] A number of entities in result set.
         * @param {number} [offset] How many entities should be skipped.
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        getPlans(code, limit, offset, options) {
            return __awaiter(this, void 0, void 0, function* () {
                const localVarAxiosArgs = yield localVarAxiosParamCreator.getPlans(code, limit, offset, options);
                return (0, common_1.createRequestFunction)(localVarAxiosArgs, axios_1.default, base_1.BASE_PATH, configuration);
            });
        },
        /**
         * This method updates a plan.
         * @summary Update plan
         * @param {string} code Code of project, where to search entities.
         * @param {number} id Identifier.
         * @param {PlanUpdate} planUpdate
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        updatePlan(code, id, planUpdate, options) {
            return __awaiter(this, void 0, void 0, function* () {
                const localVarAxiosArgs = yield localVarAxiosParamCreator.updatePlan(code, id, planUpdate, options);
                return (0, common_1.createRequestFunction)(localVarAxiosArgs, axios_1.default, base_1.BASE_PATH, configuration);
            });
        },
    };
};
exports.PlansApiFp = PlansApiFp;
/**
 * PlansApi - factory interface
 * @export
 */
const PlansApiFactory = function (configuration, basePath, axios) {
    const localVarFp = (0, exports.PlansApiFp)(configuration);
    return {
        /**
         * This method allows to create a plan in selected project.
         * @summary Create a new plan
         * @param {string} code Code of project, where to search entities.
         * @param {PlanCreate} planCreate
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        createPlan(code, planCreate, options) {
            return localVarFp.createPlan(code, planCreate, options).then((request) => request(axios, basePath));
        },
        /**
         * This method completely deletes a plan from repository.
         * @summary Delete plan
         * @param {string} code Code of project, where to search entities.
         * @param {number} id Identifier.
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        deletePlan(code, id, options) {
            return localVarFp.deletePlan(code, id, options).then((request) => request(axios, basePath));
        },
        /**
         * This method allows to retrieve a specific plan.
         * @summary Get a specific plan
         * @param {string} code Code of project, where to search entities.
         * @param {number} id Identifier.
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        getPlan(code, id, options) {
            return localVarFp.getPlan(code, id, options).then((request) => request(axios, basePath));
        },
        /**
         * This method allows to retrieve all plans stored in selected project.
         * @summary Get all plans
         * @param {string} code Code of project, where to search entities.
         * @param {number} [limit] A number of entities in result set.
         * @param {number} [offset] How many entities should be skipped.
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        getPlans(code, limit, offset, options) {
            return localVarFp.getPlans(code, limit, offset, options).then((request) => request(axios, basePath));
        },
        /**
         * This method updates a plan.
         * @summary Update plan
         * @param {string} code Code of project, where to search entities.
         * @param {number} id Identifier.
         * @param {PlanUpdate} planUpdate
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        updatePlan(code, id, planUpdate, options) {
            return localVarFp.updatePlan(code, id, planUpdate, options).then((request) => request(axios, basePath));
        },
    };
};
exports.PlansApiFactory = PlansApiFactory;
/**
 * PlansApi - object-oriented interface
 * @export
 * @class PlansApi
 * @extends {BaseAPI}
 */
class PlansApi extends base_1.BaseAPI {
    /**
     * This method allows to create a plan in selected project.
     * @summary Create a new plan
     * @param {string} code Code of project, where to search entities.
     * @param {PlanCreate} planCreate
     * @param {*} [options] Override http request option.
     * @throws {RequiredError}
     * @memberof PlansApi
     */
    createPlan(code, planCreate, options) {
        return (0, exports.PlansApiFp)(this.configuration).createPlan(code, planCreate, options).then((request) => request(this.axios, this.basePath));
    }
    /**
     * This method completely deletes a plan from repository.
     * @summary Delete plan
     * @param {string} code Code of project, where to search entities.
     * @param {number} id Identifier.
     * @param {*} [options] Override http request option.
     * @throws {RequiredError}
     * @memberof PlansApi
     */
    deletePlan(code, id, options) {
        return (0, exports.PlansApiFp)(this.configuration).deletePlan(code, id, options).then((request) => request(this.axios, this.basePath));
    }
    /**
     * This method allows to retrieve a specific plan.
     * @summary Get a specific plan
     * @param {string} code Code of project, where to search entities.
     * @param {number} id Identifier.
     * @param {*} [options] Override http request option.
     * @throws {RequiredError}
     * @memberof PlansApi
     */
    getPlan(code, id, options) {
        return (0, exports.PlansApiFp)(this.configuration).getPlan(code, id, options).then((request) => request(this.axios, this.basePath));
    }
    /**
     * This method allows to retrieve all plans stored in selected project.
     * @summary Get all plans
     * @param {string} code Code of project, where to search entities.
     * @param {number} [limit] A number of entities in result set.
     * @param {number} [offset] How many entities should be skipped.
     * @param {*} [options] Override http request option.
     * @throws {RequiredError}
     * @memberof PlansApi
     */
    getPlans(code, limit, offset, options) {
        return (0, exports.PlansApiFp)(this.configuration).getPlans(code, limit, offset, options).then((request) => request(this.axios, this.basePath));
    }
    /**
     * This method updates a plan.
     * @summary Update plan
     * @param {string} code Code of project, where to search entities.
     * @param {number} id Identifier.
     * @param {PlanUpdate} planUpdate
     * @param {*} [options] Override http request option.
     * @throws {RequiredError}
     * @memberof PlansApi
     */
    updatePlan(code, id, planUpdate, options) {
        return (0, exports.PlansApiFp)(this.configuration).updatePlan(code, id, planUpdate, options).then((request) => request(this.axios, this.basePath));
    }
}
exports.PlansApi = PlansApi;
