// qase-api-client/dist/api/shared-parameters-api.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Qase.io TestOps API v1
 * Qase TestOps API v1 Specification.
 *
 * The version of the OpenAPI document: 1.0.0
 * Contact: support@qase.io
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SharedParametersApi = exports.SharedParametersApiFactory = exports.SharedParametersApiFp = exports.SharedParametersApiAxiosParamCreator = void 0;
const axios_1 = require("axios");
// Some imports not used depending on template conditions
// @ts-ignore
const common_1 = require("../common");
// @ts-ignore
const base_1 = require("../base");
/**
 * SharedParametersApi - axios parameter creator
 * @export
 */
const SharedParametersApiAxiosParamCreator = function (configuration) {
    return {
        /**
         *
         * @summary Create a new shared parameter
         * @param {SharedParameterCreate} sharedParameterCreate
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        createSharedParameter: (sharedParameterCreate_1, ...args_1) => __awaiter(this, [sharedParameterCreate_1, ...args_1], void 0, function* (sharedParameterCreate, options = {}) {
            // verify required parameter 'sharedParameterCreate' is not null or undefined
            (0, common_1.assertParamExists)('createSharedParameter', 'sharedParameterCreate', sharedParameterCreate);
            const localVarPath = `/shared_parameter`;
            // use dummy base URL string because the URL constructor only accepts absolute URLs.
            const localVarUrlObj = new URL(localVarPath, common_1.DUMMY_BASE_URL);
            let baseOptions;
            if (configuration) {
                baseOptions = configuration.baseOptions;
            }
            const localVarRequestOptions = Object.assign(Object.assign({ method: 'POST' }, baseOptions), options);
            const localVarHeaderParameter = {};
            const localVarQueryParameter = {};
            // authentication TokenAuth required
            yield (0, common_1.setApiKeyToObject)(localVarHeaderParameter, "Token", configuration);
            localVarHeaderParameter['Content-Type'] = 'application/json';
            (0, common_1.setSearchParams)(localVarUrlObj, localVarQueryParameter);
            let headersFromBaseOptions = baseOptions && baseOptions.headers ? baseOptions.headers : {};
            localVarRequestOptions.headers = Object.assign(Object.assign(Object.assign({}, localVarHeaderParameter), headersFromBaseOptions), options.headers);
            localVarRequestOptions.data = (0, common_1.serializeDataIfNeeded)(sharedParameterCreate, localVarRequestOptions, configuration);
            return {
                url: (0, common_1.toPathString)(localVarUrlObj),
                options: localVarRequestOptions,
            };
        }),
        /**
         * Delete shared parameter along with all its usages in test cases and reviews.
         * @summary Delete shared parameter
         * @param {string} id Identifier.
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        deleteSharedParameter: (id_1, ...args_1) => __awaiter(this, [id_1, ...args_1], void 0, function* (id, options = {}) {
            // verify required parameter 'id' is not null or undefined
            (0, common_1.assertParamExists)('deleteSharedParameter', 'id', id);
            const localVarPath = `/shared_parameter/{id}`
                .replace(`{${"id"}}`, encodeURIComponent(String(id)));
            // use dummy base URL string because the URL constructor only accepts absolute URLs.
            const localVarUrlObj = new URL(localVarPath, common_1.DUMMY_BASE_URL);
            let baseOptions;
            if (configuration) {
                baseOptions = configuration.baseOptions;
            }
            const localVarRequestOptions = Object.assign(Object.assign({ method: 'DELETE' }, baseOptions), options);
            const localVarHeaderParameter = {};
            const localVarQueryParameter = {};
            // authentication TokenAuth required
            yield (0, common_1.setApiKeyToObject)(localVarHeaderParameter, "Token", configuration);
            (0, common_1.setSearchParams)(localVarUrlObj, localVarQueryParameter);
            let headersFromBaseOptions = baseOptions && baseOptions.headers ? baseOptions.headers : {};
            localVarRequestOptions.headers = Object.assign(Object.assign(Object.assign({}, localVarHeaderParameter), headersFromBaseOptions), options.headers);
            return {
                url: (0, common_1.toPathString)(localVarUrlObj),
                options: localVarRequestOptions,
            };
        }),
        /**
         *
         * @summary Get a specific shared parameter
         * @param {string} id Identifier.
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        getSharedParameter: (id_1, ...args_1) => __awaiter(this, [id_1, ...args_1], void 0, function* (id, options = {}) {
            // verify required parameter 'id' is not null or undefined
            (0, common_1.assertParamExists)('getSharedParameter', 'id', id);
            const localVarPath = `/shared_parameter/{id}`
                .replace(`{${"id"}}`, encodeURIComponent(String(id)));
            // use dummy base URL string because the URL constructor only accepts absolute URLs.
            const localVarUrlObj = new URL(localVarPath, common_1.DUMMY_BASE_URL);
            let baseOptions;
            if (configuration) {
                baseOptions = configuration.baseOptions;
            }
            const localVarRequestOptions = Object.assign(Object.assign({ method: 'GET' }, baseOptions), options);
            const localVarHeaderParameter = {};
            const localVarQueryParameter = {};
            // authentication TokenAuth required
            yield (0, common_1.setApiKeyToObject)(localVarHeaderParameter, "Token", configuration);
            (0, common_1.setSearchParams)(localVarUrlObj, localVarQueryParameter);
            let headersFromBaseOptions = baseOptions && baseOptions.headers ? baseOptions.headers : {};
            localVarRequestOptions.headers = Object.assign(Object.assign(Object.assign({}, localVarHeaderParameter), headersFromBaseOptions), options.headers);
            return {
                url: (0, common_1.toPathString)(localVarUrlObj),
                options: localVarRequestOptions,
            };
        }),
        /**
         *
         * @summary Get all shared parameters
         * @param {number} [limit] A number of entities in result set.
         * @param {number} [offset] How many entities should be skipped.
         * @param {string} [filtersSearch]
         * @param {'single' | 'group'} [filtersType]
         * @param {Array<string>} [filtersProjectCodes]
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        getSharedParameters: (limit_1, offset_1, filtersSearch_1, filtersType_1, filtersProjectCodes_1, ...args_1) => __awaiter(this, [limit_1, offset_1, filtersSearch_1, filtersType_1, filtersProjectCodes_1, ...args_1], void 0, function* (limit, offset, filtersSearch, filtersType, filtersProjectCodes, options = {}) {
            const localVarPath = `/shared_parameter`;
            // use dummy base URL string because the URL constructor only accepts absolute URLs.
            const localVarUrlObj = new URL(localVarPath, common_1.DUMMY_BASE_URL);
            let baseOptions;
            if (configuration) {
                baseOptions = configuration.baseOptions;
            }
            const localVarRequestOptions = Object.assign(Object.assign({ method: 'GET' }, baseOptions), options);
            const localVarHeaderParameter = {};
            const localVarQueryParameter = {};
            // authentication TokenAuth required
            yield (0, common_1.setApiKeyToObject)(localVarHeaderParameter, "Token", configuration);
            if (limit !== undefined) {
                localVarQueryParameter['limit'] = limit;
            }
            if (offset !== undefined) {
                localVarQueryParameter['offset'] = offset;
            }
            if (filtersSearch !== undefined) {
                localVarQueryParameter['filters[search]'] = filtersSearch;
            }
            if (filtersType !== undefined) {
                localVarQueryParameter['filters[type]'] = filtersType;
            }
            if (filtersProjectCodes) {
                localVarQueryParameter['filters[project_codes][]'] = filtersProjectCodes.join(base_1.COLLECTION_FORMATS.csv);
            }
            (0, common_1.setSearchParams)(localVarUrlObj, localVarQueryParameter);
            let headersFromBaseOptions = baseOptions && baseOptions.headers ? baseOptions.headers : {};
            localVarRequestOptions.headers = Object.assign(Object.assign(Object.assign({}, localVarHeaderParameter), headersFromBaseOptions), options.headers);
            return {
                url: (0, common_1.toPathString)(localVarUrlObj),
                options: localVarRequestOptions,
            };
        }),
        /**
         *
         * @summary Update shared parameter
         * @param {string} id Identifier.
         * @param {SharedParameterUpdate} sharedParameterUpdate
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        updateSharedParameter: (id_1, sharedParameterUpdate_1, ...args_1) => __awaiter(this, [id_1, sharedParameterUpdate_1, ...args_1], void 0, function* (id, sharedParameterUpdate, options = {}) {
            // verify required parameter 'id' is not null or undefined
            (0, common_1.assertParamExists)('updateSharedParameter', 'id', id);
            // verify required parameter 'sharedParameterUpdate' is not null or undefined
            (0, common_1.assertParamExists)('updateSharedParameter', 'sharedParameterUpdate', sharedParameterUpdate);
            const localVarPath = `/shared_parameter/{id}`
                .replace(`{${"id"}}`, encodeURIComponent(String(id)));
            // use dummy base URL string because the URL constructor only accepts absolute URLs.
            const localVarUrlObj = new URL(localVarPath, common_1.DUMMY_BASE_URL);
            let baseOptions;
            if (configuration) {
                baseOptions = configuration.baseOptions;
            }
            const localVarRequestOptions = Object.assign(Object.assign({ method: 'PATCH' }, baseOptions), options);
            const localVarHeaderParameter = {};
            const localVarQueryParameter = {};
            // authentication TokenAuth required
            yield (0, common_1.setApiKeyToObject)(localVarHeaderParameter, "Token", configuration);
            localVarHeaderParameter['Content-Type'] = 'application/json';
            (0, common_1.setSearchParams)(localVarUrlObj, localVarQueryParameter);
            let headersFromBaseOptions = baseOptions && baseOptions.headers ? baseOptions.headers : {};
            localVarRequestOptions.headers = Object.assign(Object.assign(Object.assign({}, localVarHeaderParameter), headersFromBaseOptions), options.headers);
            localVarRequestOptions.data = (0, common_1.serializeDataIfNeeded)(sharedParameterUpdate, localVarRequestOptions, configuration);
            return {
                url: (0, common_1.toPathString)(localVarUrlObj),
                options: localVarRequestOptions,
            };
        }),
    };
};
exports.SharedParametersApiAxiosParamCreator = SharedParametersApiAxiosParamCreator;
/**
 * SharedParametersApi - functional programming interface
 * @export
 */
const SharedParametersApiFp = function (configuration) {
    const localVarAxiosParamCreator = (0, exports.SharedParametersApiAxiosParamCreator)(configuration);
    return {
        /**
         *
         * @summary Create a new shared parameter
         * @param {SharedParameterCreate} sharedParameterCreate
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        createSharedParameter(sharedParameterCreate, options) {
            return __awaiter(this, void 0, void 0, function* () {
                const localVarAxiosArgs = yield localVarAxiosParamCreator.createSharedParameter(sharedParameterCreate, options);
                return (0, common_1.createRequestFunction)(localVarAxiosArgs, axios_1.default, base_1.BASE_PATH, configuration);
            });
        },
        /**
         * Delete shared parameter along with all its usages in test cases and reviews.
         * @summary Delete shared parameter
         * @param {string} id Identifier.
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        deleteSharedParameter(id, options) {
            return __awaiter(this, void 0, void 0, function* () {
                const localVarAxiosArgs = yield localVarAxiosParamCreator.deleteSharedParameter(id, options);
                return (0, common_1.createRequestFunction)(localVarAxiosArgs, axios_1.default, base_1.BASE_PATH, configuration);
            });
        },
        /**
         *
         * @summary Get a specific shared parameter
         * @param {string} id Identifier.
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        getSharedParameter(id, options) {
            return __awaiter(this, void 0, void 0, function* () {
                const localVarAxiosArgs = yield localVarAxiosParamCreator.getSharedParameter(id, options);
                return (0, common_1.createRequestFunction)(localVarAxiosArgs, axios_1.default, base_1.BASE_PATH, configuration);
            });
        },
        /**
         *
         * @summary Get all shared parameters
         * @param {number} [limit] A number of entities in result set.
         * @param {number} [offset] How many entities should be skipped.
         * @param {string} [filtersSearch]
         * @param {'single' | 'group'} [filtersType]
         * @param {Array<string>} [filtersProjectCodes]
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        getSharedParameters(limit, offset, filtersSearch, filtersType, filtersProjectCodes, options) {
            return __awaiter(this, void 0, void 0, function* () {
                const localVarAxiosArgs = yield localVarAxiosParamCreator.getSharedParameters(limit, offset, filtersSearch, filtersType, filtersProjectCodes, options);
                return (0, common_1.createRequestFunction)(localVarAxiosArgs, axios_1.default, base_1.BASE_PATH, configuration);
            });
        },
        /**
         *
         * @summary Update shared parameter
         * @param {string} id Identifier.
         * @param {SharedParameterUpdate} sharedParameterUpdate
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        updateSharedParameter(id, sharedParameterUpdate, options) {
            return __awaiter(this, void 0, void 0, function* () {
                const localVarAxiosArgs = yield localVarAxiosParamCreator.updateSharedParameter(id, sharedParameterUpdate, options);
                return (0, common_1.createRequestFunction)(localVarAxiosArgs, axios_1.default, base_1.BASE_PATH, configuration);
            });
        },
    };
};
exports.SharedParametersApiFp = SharedParametersApiFp;
/**
 * SharedParametersApi - factory interface
 * @export
 */
const SharedParametersApiFactory = function (configuration, basePath, axios) {
    const localVarFp = (0, exports.SharedParametersApiFp)(configuration);
    return {
        /**
         *
         * @summary Create a new shared parameter
         * @param {SharedParameterCreate} sharedParameterCreate
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        createSharedParameter(sharedParameterCreate, options) {
            return localVarFp.createSharedParameter(sharedParameterCreate, options).then((request) => request(axios, basePath));
        },
        /**
         * Delete shared parameter along with all its usages in test cases and reviews.
         * @summary Delete shared parameter
         * @param {string} id Identifier.
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        deleteSharedParameter(id, options) {
            return localVarFp.deleteSharedParameter(id, options).then((request) => request(axios, basePath));
        },
        /**
         *
         * @summary Get a specific shared parameter
         * @param {string} id Identifier.
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        getSharedParameter(id, options) {
            return localVarFp.getSharedParameter(id, options).then((request) => request(axios, basePath));
        },
        /**
         *
         * @summary Get all shared parameters
         * @param {number} [limit] A number of entities in result set.
         * @param {number} [offset] How many entities should be skipped.
         * @param {string} [filtersSearch]
         * @param {'single' | 'group'} [filtersType]
         * @param {Array<string>} [filtersProjectCodes]
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        getSharedParameters(limit, offset, filtersSearch, filtersType, filtersProjectCodes, options) {
            return localVarFp.getSharedParameters(limit, offset, filtersSearch, filtersType, filtersProjectCodes, options).then((request) => request(axios, basePath));
        },
        /**
         *
         * @summary Update shared parameter
         * @param {string} id Identifier.
         * @param {SharedParameterUpdate} sharedParameterUpdate
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        updateSharedParameter(id, sharedParameterUpdate, options) {
            return localVarFp.updateSharedParameter(id, sharedParameterUpdate, options).then((request) => request(axios, basePath));
        },
    };
};
exports.SharedParametersApiFactory = SharedParametersApiFactory;
/**
 * SharedParametersApi - object-oriented interface
 * @export
 * @class SharedParametersApi
 * @extends {BaseAPI}
 */
class SharedParametersApi extends base_1.BaseAPI {
    /**
     *
     * @summary Create a new shared parameter
     * @param {SharedParameterCreate} sharedParameterCreate
     * @param {*} [options] Override http request option.
     * @throws {RequiredError}
     * @memberof SharedParametersApi
     */
    createSharedParameter(sharedParameterCreate, options) {
        return (0, exports.SharedParametersApiFp)(this.configuration).createSharedParameter(sharedParameterCreate, options).then((request) => request(this.axios, this.basePath));
    }
    /**
     * Delete shared parameter along with all its usages in test cases and reviews.
     * @summary Delete shared parameter
     * @param {string} id Identifier.
     * @param {*} [options] Override http request option.
     * @throws {RequiredError}
     * @memberof SharedParametersApi
     */
    deleteSharedParameter(id, options) {
        return (0, exports.SharedParametersApiFp)(this.configuration).deleteSharedParameter(id, options).then((request) => request(this.axios, this.basePath));
    }
    /**
     *
     * @summary Get a specific shared parameter
     * @param {string} id Identifier.
     * @param {*} [options] Override http request option.
     * @throws {RequiredError}
     * @memberof SharedParametersApi
     */
    getSharedParameter(id, options) {
        return (0, exports.SharedParametersApiFp)(this.configuration).getSharedParameter(id, options).then((request) => request(this.axios, this.basePath));
    }
    /**
     *
     * @summary Get all shared parameters
     * @param {number} [limit] A number of entities in result set.
     * @param {number} [offset] How many entities should be skipped.
     * @param {string} [filtersSearch]
     * @param {'single' | 'group'} [filtersType]
     * @param {Array<string>} [filtersProjectCodes]
     * @param {*} [options] Override http request option.
     * @throws {RequiredError}
     * @memberof SharedParametersApi
     */
    getSharedParameters(limit, offset, filtersSearch, filtersType, filtersProjectCodes, options) {
        return (0, exports.SharedParametersApiFp)(this.configuration).getSharedParameters(limit, offset, filtersSearch, filtersType, filtersProjectCodes, options).then((request) => request(this.axios, this.basePath));
    }
    /**
     *
     * @summary Update shared parameter
     * @param {string} id Identifier.
     * @param {SharedParameterUpdate} sharedParameterUpdate
     * @param {*} [options] Override http request option.
     * @throws {RequiredError}
     * @memberof SharedParametersApi
     */
    updateSharedParameter(id, sharedParameterUpdate, options) {
        return (0, exports.SharedParametersApiFp)(this.configuration).updateSharedParameter(id, sharedParameterUpdate, options).then((request) => request(this.axios, this.basePath));
    }
}
exports.SharedParametersApi = SharedParametersApi;
