// qase-api-client/dist/api/shared-steps-api.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Qase.io TestOps API v1
 * Qase TestOps API v1 Specification.
 *
 * The version of the OpenAPI document: 1.0.0
 * Contact: support@qase.io
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SharedStepsApi = exports.SharedStepsApiFactory = exports.SharedStepsApiFp = exports.SharedStepsApiAxiosParamCreator = void 0;
const axios_1 = require("axios");
// Some imports not used depending on template conditions
// @ts-ignore
const common_1 = require("../common");
// @ts-ignore
const base_1 = require("../base");
/**
 * SharedStepsApi - axios parameter creator
 * @export
 */
const SharedStepsApiAxiosParamCreator = function (configuration) {
    return {
        /**
         * This method allows to create a shared step in selected project.
         * @summary Create a new shared step
         * @param {string} code Code of project, where to search entities.
         * @param {SharedStepCreate} sharedStepCreate
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        createSharedStep: (code_1, sharedStepCreate_1, ...args_1) => __awaiter(this, [code_1, sharedStepCreate_1, ...args_1], void 0, function* (code, sharedStepCreate, options = {}) {
            // verify required parameter 'code' is not null or undefined
            (0, common_1.assertParamExists)('createSharedStep', 'code', code);
            // verify required parameter 'sharedStepCreate' is not null or undefined
            (0, common_1.assertParamExists)('createSharedStep', 'sharedStepCreate', sharedStepCreate);
            const localVarPath = `/shared_step/{code}`
                .replace(`{${"code"}}`, encodeURIComponent(String(code)));
            // use dummy base URL string because the URL constructor only accepts absolute URLs.
            const localVarUrlObj = new URL(localVarPath, common_1.DUMMY_BASE_URL);
            let baseOptions;
            if (configuration) {
                baseOptions = configuration.baseOptions;
            }
            const localVarRequestOptions = Object.assign(Object.assign({ method: 'POST' }, baseOptions), options);
            const localVarHeaderParameter = {};
            const localVarQueryParameter = {};
            // authentication TokenAuth required
            yield (0, common_1.setApiKeyToObject)(localVarHeaderParameter, "Token", configuration);
            localVarHeaderParameter['Content-Type'] = 'application/json';
            (0, common_1.setSearchParams)(localVarUrlObj, localVarQueryParameter);
            let headersFromBaseOptions = baseOptions && baseOptions.headers ? baseOptions.headers : {};
            localVarRequestOptions.headers = Object.assign(Object.assign(Object.assign({}, localVarHeaderParameter), headersFromBaseOptions), options.headers);
            localVarRequestOptions.data = (0, common_1.serializeDataIfNeeded)(sharedStepCreate, localVarRequestOptions, configuration);
            return {
                url: (0, common_1.toPathString)(localVarUrlObj),
                options: localVarRequestOptions,
            };
        }),
        /**
         * This method completely deletes a shared step from repository.
         * @summary Delete shared step
         * @param {string} code Code of project, where to search entities.
         * @param {string} hash Hash.
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        deleteSharedStep: (code_1, hash_1, ...args_1) => __awaiter(this, [code_1, hash_1, ...args_1], void 0, function* (code, hash, options = {}) {
            // verify required parameter 'code' is not null or undefined
            (0, common_1.assertParamExists)('deleteSharedStep', 'code', code);
            // verify required parameter 'hash' is not null or undefined
            (0, common_1.assertParamExists)('deleteSharedStep', 'hash', hash);
            const localVarPath = `/shared_step/{code}/{hash}`
                .replace(`{${"code"}}`, encodeURIComponent(String(code)))
                .replace(`{${"hash"}}`, encodeURIComponent(String(hash)));
            // use dummy base URL string because the URL constructor only accepts absolute URLs.
            const localVarUrlObj = new URL(localVarPath, common_1.DUMMY_BASE_URL);
            let baseOptions;
            if (configuration) {
                baseOptions = configuration.baseOptions;
            }
            const localVarRequestOptions = Object.assign(Object.assign({ method: 'DELETE' }, baseOptions), options);
            const localVarHeaderParameter = {};
            const localVarQueryParameter = {};
            // authentication TokenAuth required
            yield (0, common_1.setApiKeyToObject)(localVarHeaderParameter, "Token", configuration);
            (0, common_1.setSearchParams)(localVarUrlObj, localVarQueryParameter);
            let headersFromBaseOptions = baseOptions && baseOptions.headers ? baseOptions.headers : {};
            localVarRequestOptions.headers = Object.assign(Object.assign(Object.assign({}, localVarHeaderParameter), headersFromBaseOptions), options.headers);
            return {
                url: (0, common_1.toPathString)(localVarUrlObj),
                options: localVarRequestOptions,
            };
        }),
        /**
         * This method allows to retrieve a specific shared step.
         * @summary Get a specific shared step
         * @param {string} code Code of project, where to search entities.
         * @param {string} hash Hash.
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        getSharedStep: (code_1, hash_1, ...args_1) => __awaiter(this, [code_1, hash_1, ...args_1], void 0, function* (code, hash, options = {}) {
            // verify required parameter 'code' is not null or undefined
            (0, common_1.assertParamExists)('getSharedStep', 'code', code);
            // verify required parameter 'hash' is not null or undefined
            (0, common_1.assertParamExists)('getSharedStep', 'hash', hash);
            const localVarPath = `/shared_step/{code}/{hash}`
                .replace(`{${"code"}}`, encodeURIComponent(String(code)))
                .replace(`{${"hash"}}`, encodeURIComponent(String(hash)));
            // use dummy base URL string because the URL constructor only accepts absolute URLs.
            const localVarUrlObj = new URL(localVarPath, common_1.DUMMY_BASE_URL);
            let baseOptions;
            if (configuration) {
                baseOptions = configuration.baseOptions;
            }
            const localVarRequestOptions = Object.assign(Object.assign({ method: 'GET' }, baseOptions), options);
            const localVarHeaderParameter = {};
            const localVarQueryParameter = {};
            // authentication TokenAuth required
            yield (0, common_1.setApiKeyToObject)(localVarHeaderParameter, "Token", configuration);
            (0, common_1.setSearchParams)(localVarUrlObj, localVarQueryParameter);
            let headersFromBaseOptions = baseOptions && baseOptions.headers ? baseOptions.headers : {};
            localVarRequestOptions.headers = Object.assign(Object.assign(Object.assign({}, localVarHeaderParameter), headersFromBaseOptions), options.headers);
            return {
                url: (0, common_1.toPathString)(localVarUrlObj),
                options: localVarRequestOptions,
            };
        }),
        /**
         * This method allows to retrieve all shared steps stored in selected project.
         * @summary Get all shared steps
         * @param {string} code Code of project, where to search entities.
         * @param {string} [search] Provide a string that will be used to search by name.
         * @param {number} [limit] A number of entities in result set.
         * @param {number} [offset] How many entities should be skipped.
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        getSharedSteps: (code_1, search_1, limit_1, offset_1, ...args_1) => __awaiter(this, [code_1, search_1, limit_1, offset_1, ...args_1], void 0, function* (code, search, limit, offset, options = {}) {
            // verify required parameter 'code' is not null or undefined
            (0, common_1.assertParamExists)('getSharedSteps', 'code', code);
            const localVarPath = `/shared_step/{code}`
                .replace(`{${"code"}}`, encodeURIComponent(String(code)));
            // use dummy base URL string because the URL constructor only accepts absolute URLs.
            const localVarUrlObj = new URL(localVarPath, common_1.DUMMY_BASE_URL);
            let baseOptions;
            if (configuration) {
                baseOptions = configuration.baseOptions;
            }
            const localVarRequestOptions = Object.assign(Object.assign({ method: 'GET' }, baseOptions), options);
            const localVarHeaderParameter = {};
            const localVarQueryParameter = {};
            // authentication TokenAuth required
            yield (0, common_1.setApiKeyToObject)(localVarHeaderParameter, "Token", configuration);
            if (search !== undefined) {
                localVarQueryParameter['search'] = search;
            }
            if (limit !== undefined) {
                localVarQueryParameter['limit'] = limit;
            }
            if (offset !== undefined) {
                localVarQueryParameter['offset'] = offset;
            }
            (0, common_1.setSearchParams)(localVarUrlObj, localVarQueryParameter);
            let headersFromBaseOptions = baseOptions && baseOptions.headers ? baseOptions.headers : {};
            localVarRequestOptions.headers = Object.assign(Object.assign(Object.assign({}, localVarHeaderParameter), headersFromBaseOptions), options.headers);
            return {
                url: (0, common_1.toPathString)(localVarUrlObj),
                options: localVarRequestOptions,
            };
        }),
        /**
         * This method updates a shared step.
         * @summary Update shared step
         * @param {string} code Code of project, where to search entities.
         * @param {string} hash Hash.
         * @param {SharedStepUpdate} sharedStepUpdate
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        updateSharedStep: (code_1, hash_1, sharedStepUpdate_1, ...args_1) => __awaiter(this, [code_1, hash_1, sharedStepUpdate_1, ...args_1], void 0, function* (code, hash, sharedStepUpdate, options = {}) {
            // verify required parameter 'code' is not null or undefined
            (0, common_1.assertParamExists)('updateSharedStep', 'code', code);
            // verify required parameter 'hash' is not null or undefined
            (0, common_1.assertParamExists)('updateSharedStep', 'hash', hash);
            // verify required parameter 'sharedStepUpdate' is not null or undefined
            (0, common_1.assertParamExists)('updateSharedStep', 'sharedStepUpdate', sharedStepUpdate);
            const localVarPath = `/shared_step/{code}/{hash}`
                .replace(`{${"code"}}`, encodeURIComponent(String(code)))
                .replace(`{${"hash"}}`, encodeURIComponent(String(hash)));
            // use dummy base URL string because the URL constructor only accepts absolute URLs.
            const localVarUrlObj = new URL(localVarPath, common_1.DUMMY_BASE_URL);
            let baseOptions;
            if (configuration) {
                baseOptions = configuration.baseOptions;
            }
            const localVarRequestOptions = Object.assign(Object.assign({ method: 'PATCH' }, baseOptions), options);
            const localVarHeaderParameter = {};
            const localVarQueryParameter = {};
            // authentication TokenAuth required
            yield (0, common_1.setApiKeyToObject)(localVarHeaderParameter, "Token", configuration);
            localVarHeaderParameter['Content-Type'] = 'application/json';
            (0, common_1.setSearchParams)(localVarUrlObj, localVarQueryParameter);
            let headersFromBaseOptions = baseOptions && baseOptions.headers ? baseOptions.headers : {};
            localVarRequestOptions.headers = Object.assign(Object.assign(Object.assign({}, localVarHeaderParameter), headersFromBaseOptions), options.headers);
            localVarRequestOptions.data = (0, common_1.serializeDataIfNeeded)(sharedStepUpdate, localVarRequestOptions, configuration);
            return {
                url: (0, common_1.toPathString)(localVarUrlObj),
                options: localVarRequestOptions,
            };
        }),
    };
};
exports.SharedStepsApiAxiosParamCreator = SharedStepsApiAxiosParamCreator;
/**
 * SharedStepsApi - functional programming interface
 * @export
 */
const SharedStepsApiFp = function (configuration) {
    const localVarAxiosParamCreator = (0, exports.SharedStepsApiAxiosParamCreator)(configuration);
    return {
        /**
         * This method allows to create a shared step in selected project.
         * @summary Create a new shared step
         * @param {string} code Code of project, where to search entities.
         * @param {SharedStepCreate} sharedStepCreate
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        createSharedStep(code, sharedStepCreate, options) {
            return __awaiter(this, void 0, void 0, function* () {
                const localVarAxiosArgs = yield localVarAxiosParamCreator.createSharedStep(code, sharedStepCreate, options);
                return (0, common_1.createRequestFunction)(localVarAxiosArgs, axios_1.default, base_1.BASE_PATH, configuration);
            });
        },
        /**
         * This method completely deletes a shared step from repository.
         * @summary Delete shared step
         * @param {string} code Code of project, where to search entities.
         * @param {string} hash Hash.
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        deleteSharedStep(code, hash, options) {
            return __awaiter(this, void 0, void 0, function* () {
                const localVarAxiosArgs = yield localVarAxiosParamCreator.deleteSharedStep(code, hash, options);
                return (0, common_1.createRequestFunction)(localVarAxiosArgs, axios_1.default, base_1.BASE_PATH, configuration);
            });
        },
        /**
         * This method allows to retrieve a specific shared step.
         * @summary Get a specific shared step
         * @param {string} code Code of project, where to search entities.
         * @param {string} hash Hash.
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        getSharedStep(code, hash, options) {
            return __awaiter(this, void 0, void 0, function* () {
                const localVarAxiosArgs = yield localVarAxiosParamCreator.getSharedStep(code, hash, options);
                return (0, common_1.createRequestFunction)(localVarAxiosArgs, axios_1.default, base_1.BASE_PATH, configuration);
            });
        },
        /**
         * This method allows to retrieve all shared steps stored in selected project.
         * @summary Get all shared steps
         * @param {string} code Code of project, where to search entities.
         * @param {string} [search] Provide a string that will be used to search by name.
         * @param {number} [limit] A number of entities in result set.
         * @param {number} [offset] How many entities should be skipped.
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        getSharedSteps(code, search, limit, offset, options) {
            return __awaiter(this, void 0, void 0, function* () {
                const localVarAxiosArgs = yield localVarAxiosParamCreator.getSharedSteps(code, search, limit, offset, options);
                return (0, common_1.createRequestFunction)(localVarAxiosArgs, axios_1.default, base_1.BASE_PATH, configuration);
            });
        },
        /**
         * This method updates a shared step.
         * @summary Update shared step
         * @param {string} code Code of project, where to search entities.
         * @param {string} hash Hash.
         * @param {SharedStepUpdate} sharedStepUpdate
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        updateSharedStep(code, hash, sharedStepUpdate, options) {
            return __awaiter(this, void 0, void 0, function* () {
                const localVarAxiosArgs = yield localVarAxiosParamCreator.updateSharedStep(code, hash, sharedStepUpdate, options);
                return (0, common_1.createRequestFunction)(localVarAxiosArgs, axios_1.default, base_1.BASE_PATH, configuration);
            });
        },
    };
};
exports.SharedStepsApiFp = SharedStepsApiFp;
/**
 * SharedStepsApi - factory interface
 * @export
 */
const SharedStepsApiFactory = function (configuration, basePath, axios) {
    const localVarFp = (0, exports.SharedStepsApiFp)(configuration);
    return {
        /**
         * This method allows to create a shared step in selected project.
         * @summary Create a new shared step
         * @param {string} code Code of project, where to search entities.
         * @param {SharedStepCreate} sharedStepCreate
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        createSharedStep(code, sharedStepCreate, options) {
            return localVarFp.createSharedStep(code, sharedStepCreate, options).then((request) => request(axios, basePath));
        },
        /**
         * This method completely deletes a shared step from repository.
         * @summary Delete shared step
         * @param {string} code Code of project, where to search entities.
         * @param {string} hash Hash.
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        deleteSharedStep(code, hash, options) {
            return localVarFp.deleteSharedStep(code, hash, options).then((request) => request(axios, basePath));
        },
        /**
         * This method allows to retrieve a specific shared step.
         * @summary Get a specific shared step
         * @param {string} code Code of project, where to search entities.
         * @param {string} hash Hash.
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        getSharedStep(code, hash, options) {
            return localVarFp.getSharedStep(code, hash, options).then((request) => request(axios, basePath));
        },
        /**
         * This method allows to retrieve all shared steps stored in selected project.
         * @summary Get all shared steps
         * @param {string} code Code of project, where to search entities.
         * @param {string} [search] Provide a string that will be used to search by name.
         * @param {number} [limit] A number of entities in result set.
         * @param {number} [offset] How many entities should be skipped.
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        getSharedSteps(code, search, limit, offset, options) {
            return localVarFp.getSharedSteps(code, search, limit, offset, options).then((request) => request(axios, basePath));
        },
        /**
         * This method updates a shared step.
         * @summary Update shared step
         * @param {string} code Code of project, where to search entities.
         * @param {string} hash Hash.
         * @param {SharedStepUpdate} sharedStepUpdate
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        updateSharedStep(code, hash, sharedStepUpdate, options) {
            return localVarFp.updateSharedStep(code, hash, sharedStepUpdate, options).then((request) => request(axios, basePath));
        },
    };
};
exports.SharedStepsApiFactory = SharedStepsApiFactory;
/**
 * SharedStepsApi - object-oriented interface
 * @export
 * @class SharedStepsApi
 * @extends {BaseAPI}
 */
class SharedStepsApi extends base_1.BaseAPI {
    /**
     * This method allows to create a shared step in selected project.
     * @summary Create a new shared step
     * @param {string} code Code of project, where to search entities.
     * @param {SharedStepCreate} sharedStepCreate
     * @param {*} [options] Override http request option.
     * @throws {RequiredError}
     * @memberof SharedStepsApi
     */
    createSharedStep(code, sharedStepCreate, options) {
        return (0, exports.SharedStepsApiFp)(this.configuration).createSharedStep(code, sharedStepCreate, options).then((request) => request(this.axios, this.basePath));
    }
    /**
     * This method completely deletes a shared step from repository.
     * @summary Delete shared step
     * @param {string} code Code of project, where to search entities.
     * @param {string} hash Hash.
     * @param {*} [options] Override http request option.
     * @throws {RequiredError}
     * @memberof SharedStepsApi
     */
    deleteSharedStep(code, hash, options) {
        return (0, exports.SharedStepsApiFp)(this.configuration).deleteSharedStep(code, hash, options).then((request) => request(this.axios, this.basePath));
    }
    /**
     * This method allows to retrieve a specific shared step.
     * @summary Get a specific shared step
     * @param {string} code Code of project, where to search entities.
     * @param {string} hash Hash.
     * @param {*} [options] Override http request option.
     * @throws {RequiredError}
     * @memberof SharedStepsApi
     */
    getSharedStep(code, hash, options) {
        return (0, exports.SharedStepsApiFp)(this.configuration).getSharedStep(code, hash, options).then((request) => request(this.axios, this.basePath));
    }
    /**
     * This method allows to retrieve all shared steps stored in selected project.
     * @summary Get all shared steps
     * @param {string} code Code of project, where to search entities.
     * @param {string} [search] Provide a string that will be used to search by name.
     * @param {number} [limit] A number of entities in result set.
     * @param {number} [offset] How many entities should be skipped.
     * @param {*} [options] Override http request option.
     * @throws {RequiredError}
     * @memberof SharedStepsApi
     */
    getSharedSteps(code, search, limit, offset, options) {
        return (0, exports.SharedStepsApiFp)(this.configuration).getSharedSteps(code, search, limit, offset, options).then((request) => request(this.axios, this.basePath));
    }
    /**
     * This method updates a shared step.
     * @summary Update shared step
     * @param {string} code Code of project, where to search entities.
     * @param {string} hash Hash.
     * @param {SharedStepUpdate} sharedStepUpdate
     * @param {*} [options] Override http request option.
     * @throws {RequiredError}
     * @memberof SharedStepsApi
     */
    updateSharedStep(code, hash, sharedStepUpdate, options) {
        return (0, exports.SharedStepsApiFp)(this.configuration).updateSharedStep(code, hash, sharedStepUpdate, options).then((request) => request(this.axios, this.basePath));
    }
}
exports.SharedStepsApi = SharedStepsApi;
