// qase-api-client/dist/api/system-fields-api.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/**
 * Qase.io TestOps API v1
 * Qase TestOps API v1 Specification.
 *
 * The version of the OpenAPI document: 1.0.0
 * Contact: support@qase.io
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SystemFieldsApi = exports.SystemFieldsApiFactory = exports.SystemFieldsApiFp = exports.SystemFieldsApiAxiosParamCreator = void 0;
const axios_1 = require("axios");
// Some imports not used depending on template conditions
// @ts-ignore
const common_1 = require("../common");
// @ts-ignore
const base_1 = require("../base");
/**
 * SystemFieldsApi - axios parameter creator
 * @export
 */
const SystemFieldsApiAxiosParamCreator = function (configuration) {
    return {
        /**
         * This method allows to retrieve all system fields.
         * @summary Get all System Fields
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        getSystemFields: (...args_1) => __awaiter(this, [...args_1], void 0, function* (options = {}) {
            const localVarPath = `/system_field`;
            // use dummy base URL string because the URL constructor only accepts absolute URLs.
            const localVarUrlObj = new URL(localVarPath, common_1.DUMMY_BASE_URL);
            let baseOptions;
            if (configuration) {
                baseOptions = configuration.baseOptions;
            }
            const localVarRequestOptions = Object.assign(Object.assign({ method: 'GET' }, baseOptions), options);
            const localVarHeaderParameter = {};
            const localVarQueryParameter = {};
            // authentication TokenAuth required
            yield (0, common_1.setApiKeyToObject)(localVarHeaderParameter, "Token", configuration);
            (0, common_1.setSearchParams)(localVarUrlObj, localVarQueryParameter);
            let headersFromBaseOptions = baseOptions && baseOptions.headers ? baseOptions.headers : {};
            localVarRequestOptions.headers = Object.assign(Object.assign(Object.assign({}, localVarHeaderParameter), headersFromBaseOptions), options.headers);
            return {
                url: (0, common_1.toPathString)(localVarUrlObj),
                options: localVarRequestOptions,
            };
        }),
    };
};
exports.SystemFieldsApiAxiosParamCreator = SystemFieldsApiAxiosParamCreator;
/**
 * SystemFieldsApi - functional programming interface
 * @export
 */
const SystemFieldsApiFp = function (configuration) {
    const localVarAxiosParamCreator = (0, exports.SystemFieldsApiAxiosParamCreator)(configuration);
    return {
        /**
         * This method allows to retrieve all system fields.
         * @summary Get all System Fields
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        getSystemFields(options) {
            return __awaiter(this, void 0, void 0, function* () {
                const localVarAxiosArgs = yield localVarAxiosParamCreator.getSystemFields(options);
                return (0, common_1.createRequestFunction)(localVarAxiosArgs, axios_1.default, base_1.BASE_PATH, configuration);
            });
        },
    };
};
exports.SystemFieldsApiFp = SystemFieldsApiFp;
/**
 * SystemFieldsApi - factory interface
 * @export
 */
const SystemFieldsApiFactory = function (configuration, basePath, axios) {
    const localVarFp = (0, exports.SystemFieldsApiFp)(configuration);
    return {
        /**
         * This method allows to retrieve all system fields.
         * @summary Get all System Fields
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        getSystemFields(options) {
            return localVarFp.getSystemFields(options).then((request) => request(axios, basePath));
        },
    };
};
exports.SystemFieldsApiFactory = SystemFieldsApiFactory;
/**
 * SystemFieldsApi - object-oriented interface
 * @export
 * @class SystemFieldsApi
 * @extends {BaseAPI}
 */
class SystemFieldsApi extends base_1.BaseAPI {
    /**
     * This method allows to retrieve all system fields.
     * @summary Get all System Fields
     * @param {*} [options] Override http request option.
     * @throws {RequiredError}
     * @memberof SystemFieldsApi
     */
    getSystemFields(options) {
        return (0, exports.SystemFieldsApiFp)(this.configuration).getSystemFields(options).then((request) => request(this.axios, this.basePath));
    }
}
exports.SystemFieldsApi = SystemFieldsApi;
